---
title: "Genome Assembly and Annotation"
author: "Harry Chown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    highlight: pygments
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction to genome assembly and annotation



## Caveats

The aim of this tutorial is teach you the process of assembling and annotating your first fungal genome. The methodology outlined below has been specifically designed to annotate the genome of *A. fumigatus* isolates. Modifications to this tutorial will enable genome analyses of alternative fungal organisms, but please do ensure to read the original software documentation to identify where those changes will be required. I will attempt to notify any specific parameters that I use for *A. fumigatus* but I may have missed some.

Additionally, this tutorial may change over time to include assembling from various sequencing platforms and incorporation of new analysis tools. However, older versions will be available through releases.

Furthermore, the code and scripts outlined in this tutorial has been design for use on ICL HPC systems. They follow a simple format of job scripts which can be deployed on this system. Installation guidance is also provided to help members of ICL install the relevant tools on this system. This may mean that this is not 100% transferable to the system you are working on. In addition to this, folder structure will be relevant to members working within Matthew Fisher's group, so check and update the relevant file/folder paths when required. Please seek guidance elsewhere if installation and running of tools fails when used on your relevant system.

## Demonstration data

### Short-read data
Paired-end short-read sequence data derived from *A. fumigatus* strain CEA10 using NextSeq 500 under the accession [SRR12849133](https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR12849133). Members of FisherLab can find data at:
```
~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_*.fastq.gz
```
### ONT long-read data
Long-read sequence data derived from *A. fumigatus* strain CEA10 using ONT under the accession [SRR28036069](https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR28036069). Members of FisherLab can find data at:
```
~/../projects/fisher-aspergillus-rawdata/live/tutorial/SRR28036069.fastq.gz
```
# Read quality control (QC)

## Long-read QC
We are going to generate a conda environment called ```ont-qc``` which contains three tools in order to assess the quality of ONT reads, trim and filter. The three tools are:

* NanoQC - visualise quality
  + Version: 0.10.0
  + [GitHub](https://github.com/wdecoster/nanoQC) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149)
* NanoFilt - filter reads
  + Version: 2.8.0
  + [GitHub](https://github.com/wdecoster/nanofilt)
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149) (same as NanoQC)
* porechop - trim reads
  + Version: 0.2.4
  + [GitHub](https://github.com/rrwick/Porechop)
  + [Paper](https://doi.org/10.1099/mgen.0.000132)

Installation:
```
conda create -n ont-qc bioconda::nanoqc bioconda::porechop bioconda::nanofilt
```
Script for ONT read QC analysis can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_qc.job
```
Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:40:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zONT_qc
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
INPUT=~/../projects/fisher-aspergillus-rawdata/live/tutorial/SRR28036069.fastq.gz
QC_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data
#--------------------------------------------
# 3. QC long-reads
#CPU Time used: 02:54:36
#CPU Percent: 975%
#Memory usage: 6260156kb
#Approx Power usage: 0.0348
#Walltime usage: 00:29:40
#--------------------------------------------
conda activate ont-qc
# Initial QC
cd $TMPDIR
mkdir -p $QC_DIR/ONT/initial_QC
nanoQC -o $QC_DIR/ONT/initial_QC $INPUT
gunzip -c $INPUT > SRR28036069.fastq
# Read trimming and filtering
porechop -i SRR28036069.fastq -t 10 --format fastq | \
NanoFilt -q 10 -l 500  > $QC_DIR/ONT/SRR28036069.trimmed.fastq
# Post-filter QC
mkdir -p $QC_DIR/ONT/post_QC
nanoQC -o $QC_DIR/ONT/post_QC $QC_DIR/ONT/SRR28036069.trimmed.fastq
```
**Lines 12 and 13** sets the desired input files and output directory.

**Lines 26 and 33** visualse QC on reads before and after filtering, respectively

**Lines 29 and 30** trims using porechop's default settings and filters (NanoFilt) based on average read quality >10 or length <500bp. 



# Assembly

## Long-read assembly

### ONT assembly

#### Initial assembly
We are going to use a single tool to generate our initial genome assembly:

* Flye - long-read assembler
  + Version: 2.9.6-b1802
  + [GitHub](https://github.com/mikolmogorov/Flye/tree/flye) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n flye bioconda::flye
```
Script for ONT read assembly can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm.job
```

Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:40:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zONT_asm
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
INPUT=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/SRR28036069.trimmed.fastq
ASM_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/flye
#--------------------------------------------
# 3. Assemble ONT reads
# CPU Time used: 01:32:43
# CPU Percent: 968%
# Memory usage: 10680116kb
# Approx Power usage: 0.0184
# Walltime usage: 00:14:19
#--------------------------------------------
mkdir $ASM_DIR
conda activate flye
# Initial assembly
flye --nano-hq $INPUT -o $ASM_DIR --threads 10
```
**Line 25** assemble long-reads

### Improvement with short-read data

Tools for alignment, indexing and polishing:

* Minimap2 - sequence alignment
  + Version: 2.30-r1287
  + [GitHub](https://github.com/lh3/minimap2) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty191)
* SAMtools - alignment manipulation
  + Version: 1.23
  + [GitHub](https://github.com/samtools/samtools) 
  + [Paper](https://doi.org/10.1093/gigascience/giab008)
* Pilon - genome polisher
  + Version: 1.24
  + [GitHub](https://github.com/broadinstitute/pilon) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n polish bioconda::minimap2 bioconda::samtools
```
Additional installation was required for Pilon. **ICL users do not need to repeat the following steps.**
```
cd ~/../projects/fisher-local-software/live
mkdir pilon
cd pilon
wget https://github.com/broadinstitute/pilon/releases/download/v1.24/pilon-1.24.jar
```

Script for polishing assemblies with short-read data can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/polish.job
```

Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=01:00:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zPolish
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/flye/assembly.fasta
R1=~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_1.fastq.gz
R2=~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_2.fastq.gz
POLISH_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish
PILON_PATH=~/../projects/fisher-local-software/live/pilon/pilon-1.24.jar
#--------------------------------------------
# 3. Generate polish rounds
#CPU Time used: 00:41:08
#CPU Percent: 168%
#Memory usage: 12210052kb
#Approx Power usage: 0.008199999999999999
#Walltime usage: 00:35:53
#--------------------------------------------
mkdir $POLISH_DIR
conda activate polish
cd $TMPDIR
cp $GENOME ./tmp.fasta
for i in $(seq 1 5); do
	mkdir pilon_tmp
	# Align sequence data to genome
	minimap2 -ax sr tmp.fasta $R1 $R2 | samtools sort -o output.sorted.bam
	samtools index output.sorted.bam
	# Use pilon to improve genome based on alignment
	java -Xmx100G -jar $PILON_PATH --genome tmp.fasta --bam output.sorted.bam \
	--outdir pilon_tmp --changes 
	# Overwrite genome 
	mv pilon_tmp/pilon.fasta tmp.fasta
	rm -rf pilon_tmp
done

cp tmp.fasta $POLISH_DIR/ont.polished.fasta

```
**Lines 30 and 40** establishes a loop to repeat the polish 5 times

**Line 33 and 34** aligns reads to the genome, sorts and index alignment file

**Lines 36 and 37** improves the assembly using the mapped data

**Lines 38 and 39** overwrites the input file with the new output in order to run polishing again

**Line 42** saves the results in the respective output directory

# Assembly QC
Tools for quality control of assemblies:

* SeqKit - basic genome statistics
  + Version: 2.12.0
  + [GitHub](https://github.com/shenwei356/seqkit) 
  + [Paper](https://doi.org/10.1002/imt2.191)
* BUSCO - gene content of near-universal single-copy orthologs
  + Version: 6.0.0
  + [Website](https://busco.ezlab.org/) 
  + [Paper]( https://doi.org/10.1093/nar/gkae987)
* QualiMap - coverage analyser
  + Version: 2.3
  + [Website](http://qualimap.conesalab.org/) 
  + [Paper]( https://doi.org/10.1093/bioinformatics/btv566)
* Minimap2, SAMtools - for read alignments to assembly, see details above

Installation:
We will also be using minimap2 and samtools which can be found in the ```polish``` Conda environment. Installation is outlined above.
Additionally, we will be running an offline BUSCO search using the appropriate database (eurotiales) for *A. fumigatus*. Details on downloading the appropriate BUSCO database can be found at: <https://busco.ezlab.org/>.
```
conda create -n asm-qc bioconda::busco=6.0.0 bioconda::sepp=4.5.5 bioconda::qualimap bioconda::seqkit
```


Note: if assembling an organism that is not part of the *Aspergilli* you may need to update the BUSCO database; if running on an alternative machine you will need to change to the appropriate download path. Additionally, read mapping is only for long-read sequence data, please see section on polishing to edit Minimap2 for short-read data.

Script for running assembly QC (derived from long-read data) can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm_qc.job
```


Usage:

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:20:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zAsmQC
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish/ont.polished.fasta
ONT=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/SRR28036069.trimmed.fastq
BUSCO_DB=~/../projects/fisher-local-software/live/busco_downloads
ASM_QC_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/asm_QC
#--------------------------------------------
# 3. Perform QC on polished assembly
#--------------------------------------------
mkdir $ASM_QC_DIR
conda activate asm-qc
# Traditional statistics
seqkit stats -a -N 50,90 $GENOME | sed 's/,//g' > $ASM_QC_DIR/basic_stats.txt
# Conserved universal orthologs
cd $TMPDIR
mkdir busco
busco -m genome -i $GENOME -o assembly --out_path busco -l eurotiales_odb10 \
--miniprot --offline \
--download_path ~/../projects/fisher-local-software/live/busco_downloads \
-f -c 10
cp ./busco/assembly/short_summary.specific.eurotiales_odb10.assembly.txt $ASM_QC_DIR/busco.txt
# Coverage 
## ONT read mapping
conda activate polish
minimap2 -a -x map-ont $GENOME $ONT | samtools sort -T tmp -o output.sorted.bam
samtools index output.sorted.bam
## Coverage calc.
conda activate asm-qc
mkdir $ASM_QC_DIR/qualimap
qualimap bamqc -nt 10 -bam output.sorted.bam -outdir $ASM_QC_DIR/qualimap \
--java-mem-size=80G
#--------------------------------------------
# 4. Summarise QC results
#--------------------------------------------
#num_seqs sum_len min_len avg_len max_len median GC% N50 N90
STATS=$(awk '{print $4,$5,$6,$7,$8,$10,$18,$20,$21'} $ASM_QC_DIR/basic_stats.txt \
| tail -n1 | sed 's/ /,/g')
#BUSCO Complete	| BUSCO Single |	BUSCO Duplicate	 | BUSCO Fragmented | 	BUSCO Missing	| BUSCO Num Genes
BUSCO=$(grep "C:" $ASM_QC_DIR/busco.txt | awk '{print $1}' | sed 's/\[/,/g' \
| sed 's/\]//g' | sed 's/[%|:|Aa-Zz]//g' | cut -d"," -f1-6)
COV=$(cat $ASM_QC_DIR/qualimap/genome_results.txt | grep "mean coverageData" \
| cut -d"=" -f2 | sed 's/ //g' | sed 's/X//g')
echo "${STATS},${COV},${BUSCO}" > $ASM_QC_DIR/overall_stats.csv
```
**Line 22** performs basic calculation of assembly statistics

**Lines 26-29** identifies the presence of single-copy conserved orthologs.
This provides an idea of genome completeness

**Lines 33-40** aligns reads to the genome (minimap2, samtools) and calculates
coverage (qualimap), providing an idea of read-depth

**Lines 45-52** extract formatted information from outputs and saves this as a 
single comma-separated line. Exact headers are described a combination of 
**Line 44 + coverage + Line 47**

# Basic gene prediction
Installation:
```
conda create -n funannotate "python>=3.6,<3.9" bioconda::funannotate bioconda::eggnog-mapper
conda activate funannotate
funannotate setup -i all -d ~/../projects/fisher-local-db/live/funannotate_db


On a local machine download GeneMark-ES/ET/EP and move it to your desired folder from http://topaz.gatech.edu/GeneMark/license_download.cgi
Once downloaded, update the Perl script shebangs
cd ~/../projects/fisher-local-software/live/genemark/gmes_linux_64_4
for file in *.pl
do
	sed -i 's|#!/usr/bin/perl|#!/usr/bin/env perl|g' $file
done

Access to dbCAN and BUSCO outgroups failed therefore we will install all of the remaining databases
funannotate setup -i merops -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i uniprot -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i pfam -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i repeats -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i go -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i mibig -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i interpro -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i gene2product -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i busco_outgroups -f -d ~/../projects/fisher-local-db/live/funannotate_db -w


cd ~/../projects/fisher-local-db/live/eggnog_db
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog.db.gz 
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog_proteins.dmnd.gz 
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog.taxa.tar.gz 

gunzip eggnog.db.gz 
gunzip eggnog_proteins.dmnd.gz 
tar -xvzf eggnog.taxa.tar.gz 
rm eggnog.taxa.tar.gz 

All Users:
  You will need to setup the funannotate databases using funannotate setup.
  The location of these databases on the file system is your decision and the
  location can be defined using the FUNANNOTATE_DB environmental variable.
  
  To set this up in your conda environment you can run the following:
    echo "export FUNANNOTATE_DB=/your/path" > /rds/general/user/hchown/home/anaconda3/envs/funannotate/etc/conda/activate.d/funannotate.sh
    echo "unset FUNANNOTATE_DB" > /rds/general/user/hchown/home/anaconda3/envs/funannotate/etc/conda/deactivate.d/funannotate.sh
  
  You can then run your database setup using funannotate:
    funannotate setup -i all
    
  Due to licensing restrictions, if you want to use GeneMark-ES/ET, you will need to install manually:
  download and follow directions at http://topaz.gatech.edu/GeneMark/license_download.cgi
  ** note you will likely need to change shebang line for all perl scripts:
    change: #!/usr/bin/perl to #!/usr/bin/env perl
     
      
Mac OSX Users:
  Augustus and Trinity cannot be properly installed via conda/bioconda at this time. However,
  they are able to be installed manually using a local copy of GCC (gcc-8 in example below).

  Install augustus using this repo:
    https://github.com/nextgenusfs/augustus
  
  To install Trinity v2.15.2, download the source code and compile using GCC/G++:
    wget https://github.com/trinityrnaseq/trinityrnaseq/releases/download/Trinity-v2.15.2/trinityrnaseq-v2.15.2.FULL.tar.gz
    tar xzvf trinityrnaseq-v2.15.2.FULL.tar.gz
    cd trinityrnaseq-v2.15.2
    make CC=gcc-8 CXX=g++-8
    echo "export TRINITY_HOME=/your/path" > /rds/general/user/hchown/home/anaconda3/envs/funannotate/etc/conda/activate.d/trinity.sh
    echo "unset TRINITY_HOME" > /rds/general/user/hchown/home/anaconda3/envs/funannotate/etc/conda/deactivate.d/trinity.sh    


Check for genemark installation
```

Funannotate
```
# As we a
funannotate predict -i genome.fa -s "Aspergillus fumigatus" -o output --protein_evidence $FUNANNOTATE_DB/uniprot_sprot.fasta
```

# Annotating methylation sites from ONT data
Installation:
```
cd ~/../projects/fisher-local-software/live
curl "https://cdn.oxfordnanoportal.com/software/analysis/dorado-1.3.0-linux-x64.tar.gz" -o dorado-1.3.0-linux-x64.tar.gz
tar -xzf dorado-1.3.0-linux-x64.tar.gz
~/../projects/fisher-local-software/live/dorado-1.3.0-linux-x64/bin/dorado --version

conda create -n ont-meth bioconda::ont-modkit 
```
Usage:
```
# Align reads then combine
~/../projects/fisher-local-software/live/dorado-1.3.0-linux-x64/bin/dorado aligner <draft.fasta> <unmapped_reads.bam> | samtools sort --threads <num_threads> > aligned_reads.bam
```