---
title: "Genome Assembly and Annotation"
author: "Harry Chown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    highlight: pygments
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction to genome assembly and annotation



## Caveats

The aim of this tutorial is teach you the process of assembling and annotating your first fungal genome. The methodology outlined below has been specifically designed to annotate the genome of *A. fumigatus* isolates. Modifications to this tutorial will enable genome analyses of alternative fungal organisms, but please do ensure to read the original software documentation to identify where those changes will be required. I will attempt to notify any specific parameters that I use for *A. fumigatus* but I may have missed some.

Additionally, this tutorial may change over time to include assembling from various sequencing platforms and incorporation of new analysis tools. However, older versions will be available through releases.

Furthermore, the code and scripts outlined in this tutorial has been design for use on ICL HPC systems. They follow a simple format of job scripts which can be deployed on this system. Installation guidance is also provided to help members of ICL install the relevant tools on this system. This may mean that this is not 100% transferable to the system you are working on. In addition to this, folder structure will be relevant to members working within Matthew Fisher's group, so check and update the relevant file/folder paths when required. Please seek guidance elsewhere if installation and running of tools fails when used on your relevant system.

## Demonstration data

### Short-read data
Paired-end short-read sequence data derived from *A. fumigatus* strain CEA10 using NextSeq 500 under the accession [SRR12849133](https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR12849133). Members of FisherLab can find data at:
```
~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_*.fastq.gz
```
### ONT long-read data
Long-read sequence data derived from *A. fumigatus* strain CEA10 using ONT under the accession [SRR28036069](https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR28036069). Members of FisherLab can find data at:
```
~/../projects/fisher-aspergillus-rawdata/live/tutorial/SRR28036069.fastq.gz
```
# Read quality control (QC)

## Short-read QC

We will be using two tools to check the quality of our reads, filter poor quality sequences and trim adapter sequences:

* FastQC - visualise quality
  + Version: 0.10.0
  + [GitHub](https://github.com/wdecoster/nanoQC) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149)
* FastP - filter and trim reads
  + Version: 2.8.0
  + [GitHub](https://github.com/wdecoster/nanofilt)
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149) (same as NanoQC)

Installation:
```
conda create -n ont-qc bioconda::nanoqc bioconda::porechop bioconda::nanofilt
```
Script for short-read QC analysis can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_qc.job
```
Usage:
```bash {.numberLines}
conda activate fastqc
fastqc -o $OUTDIR $R1
fastqc -o $OUTDIR $R2

# Trimming and quality filtering - fastp
conda activate fastp
fastp -i $R1  -I $R2  -o $OUTDIR/${PREFIX}_1.trimmed.fastq.gz  -O $OUTDIR/${PREFIX}_2.trimmed.fastq.gz


# Re-assess trimmed quality - fastqc
conda activate fastqc
fastqc -o $OUTDIR $OUTDIR/${PREFIX}_1.trimmed.fastq.gz 
fastqc -o $OUTDIR $OUTDIR/${PREFIX}_2.trimmed.fastq.gz
```


## Long-read QC
We are going to generate a conda environment called ```ont-qc``` which contains three tools in order to assess the quality of ONT reads, trim and filter. The three tools are:

* NanoQC - visualise quality
  + Version: 0.10.0
  + [GitHub](https://github.com/wdecoster/nanoQC) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149)
* NanoFilt - filter reads
  + Version: 2.8.0
  + [GitHub](https://github.com/wdecoster/nanofilt)
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149) (same as NanoQC)
* porechop - trim reads
  + Version: 0.2.4
  + [GitHub](https://github.com/rrwick/Porechop)
  + [Paper](https://doi.org/10.1099/mgen.0.000132)

Installation:
```
conda create -n ont-qc bioconda::nanoqc bioconda::porechop bioconda::nanofilt
```
Script for ONT read QC analysis can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_qc.job
```
Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:40:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zONT_qc
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
INPUT=~/../projects/fisher-aspergillus-rawdata/live/tutorial/SRR28036069.fastq.gz
QC_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data
#--------------------------------------------
# 3. QC long-reads
#CPU Time used: 02:54:36
#CPU Percent: 975%
#Memory usage: 6260156kb
#Approx Power usage: 0.0348
#Walltime usage: 00:29:40
#--------------------------------------------
conda activate ont-qc
# Initial QC
cd $TMPDIR
mkdir -p $QC_DIR/ONT/initial_QC
nanoQC -o $QC_DIR/ONT/initial_QC $INPUT
gunzip -c $INPUT > SRR28036069.fastq
# Read trimming and filtering
porechop -i SRR28036069.fastq -t 10 --format fastq | \
NanoFilt -q 10 -l 500  > $QC_DIR/ONT/SRR28036069.trimmed.fastq
# Post-filter QC
mkdir -p $QC_DIR/ONT/post_QC
nanoQC -o $QC_DIR/ONT/post_QC $QC_DIR/ONT/SRR28036069.trimmed.fastq
```
**Lines 12 and 13** sets the desired input files and output directory.

**Lines 26 and 33** visualse QC on reads before and after filtering, respectively

**Lines 29 and 30** trims using porechop's default settings and filters (NanoFilt) based on average read quality >10 or length <500bp. 

# Removing mitochondrial sequences from short-reads
Mitochondrial sequences can impede the assembly process, therefore you can remove these prior to assembly (alternatively, you can assemble first and then remove).
Tools for mitochondrial sequence extraction:

* GetOrganelle - organelle identification
  + Version: 2.9.6-b1802
  + [GitHub](https://github.com/mikolmogorov/Flye/tree/flye) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n flye bioconda::flye
```
Script for ONT read assembly can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm.job
```

Usage:
```bash {.numberLines}
conda activate getorganelle
mkdir -p $OUTDIR/mitochondria
get_organelle_config.py --add fungus_mt
get_organelle_from_reads.py -1 $OUTDIR/${PREFIX}_1.trimmed.fastq.gz  -2 $OUTDIR/${PREFIX}_2.trimmed.fastq.gz  -t 1 -k 21,45,65,85,105 -o $OUTDIR/mitochondria -F fungus_mt -R 10 -s /rds/general/user/hchown/projects/fisher-aspergillus-reference/live/NC_017016.1_mito.fa --overwrite --reduce-reads-for-coverage inf --max-reads inf

# Read mapping, see polish


conda activate samtools
samtools view -b -f 4 $PREFIX.mit.sorted.bam > $PREFIX.mit.unmapped.bam
samtools sort -n -o $PREFIX.nucl.qsort.bam $PREFIX.mit.unmapped.bam
# Convert to fastq
conda activate bedtools
bedtools bamtofastq -i $PREFIX.nucl.qsort.bam -fq "${PREFIX}_1.nuclear.fastq" -fq2 "${PREFIX}_2.nuclear.fastq"
gzip "${PREFIX}_1.nuclear.fastq" "${PREFIX}_2.nuclear.fastq"
mv *.nuclear.fastq.gz $OUTDIR
```

# Assembly

## Short-read assembly



## Long-read assembly

### ONT assembly

#### Initial assembly
We are going to use a single tool to generate our initial genome assembly:

* Flye - long-read assembler
  + Version: 2.9.6-b1802
  + [GitHub](https://github.com/mikolmogorov/Flye/tree/flye) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n flye bioconda::flye
```
Script for ONT read assembly can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm.job
```

Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:40:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zONT_asm
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
INPUT=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/SRR28036069.trimmed.fastq
ASM_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/flye
#--------------------------------------------
# 3. Assemble ONT reads
# CPU Time used: 01:32:43
# CPU Percent: 968%
# Memory usage: 10680116kb
# Approx Power usage: 0.0184
# Walltime usage: 00:14:19
#--------------------------------------------
mkdir $ASM_DIR
conda activate flye
# Initial assembly
flye --nano-hq $INPUT -o $ASM_DIR --threads 10
```
**Line 25** assemble long-reads

### Improvement with short-read data

Tools for alignment, indexing and polishing:

* Minimap2 - sequence alignment
  + Version: 2.30-r1287
  + [GitHub](https://github.com/lh3/minimap2) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty191)
* SAMtools - alignment manipulation
  + Version: 1.23
  + [GitHub](https://github.com/samtools/samtools) 
  + [Paper](https://doi.org/10.1093/gigascience/giab008)
* Pilon - genome polisher
  + Version: 1.24
  + [GitHub](https://github.com/broadinstitute/pilon) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n polish bioconda::minimap2 bioconda::samtools
```
Additional installation was required for Pilon. **Fisher Lab users do not need to repeat the following steps.**
```
cd ~/../projects/fisher-local-software/live
mkdir pilon
cd pilon
wget https://github.com/broadinstitute/pilon/releases/download/v1.24/pilon-1.24.jar
```

Script for polishing assemblies with short-read data can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/polish.job
```

Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=01:00:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zPolish
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/flye/assembly.fasta
R1=~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_1.fastq.gz
R2=~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_2.fastq.gz
POLISH_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish
PILON_PATH=~/../projects/fisher-local-software/live/pilon/pilon-1.24.jar
#--------------------------------------------
# 3. Generate polish rounds
#CPU Time used: 00:41:08
#CPU Percent: 168%
#Memory usage: 12210052kb
#Approx Power usage: 0.008199999999999999
#Walltime usage: 00:35:53
#--------------------------------------------
mkdir $POLISH_DIR
conda activate polish
cd $TMPDIR
cp $GENOME ./tmp.fasta
for i in $(seq 1 5); do
	mkdir pilon_tmp
	# Align sequence data to genome
	minimap2 -ax sr tmp.fasta $R1 $R2 | samtools sort -o output.sorted.bam
	samtools index output.sorted.bam
	# Use pilon to improve genome based on alignment
	java -Xmx100G -jar $PILON_PATH --genome tmp.fasta --bam output.sorted.bam \
	--outdir pilon_tmp --changes 
	# Overwrite genome 
	mv pilon_tmp/pilon.fasta tmp.fasta
	rm -rf pilon_tmp
done

cp tmp.fasta $POLISH_DIR/ont.polished.fasta

```
**Lines 30 and 40** establishes a loop to repeat the polish 5 times

**Line 33 and 34** aligns reads to the genome, sorts and index alignment file

**Lines 36 and 37** improves the assembly using the mapped data

**Lines 38 and 39** overwrites the input file with the new output in order to run polishing again

**Line 42** saves the results in the respective output directory

# Assembly QC
Tools for quality control of assemblies:

* SeqKit - basic genome statistics
  + Version: 2.12.0
  + [GitHub](https://github.com/shenwei356/seqkit) 
  + [Paper](https://doi.org/10.1002/imt2.191)
* BUSCO - gene content of near-universal single-copy orthologs
  + Version: 6.0.0
  + [Website](https://busco.ezlab.org/) 
  + [Paper]( https://doi.org/10.1093/nar/gkae987)
* QualiMap - coverage analyser
  + Version: 2.3
  + [Website](http://qualimap.conesalab.org/) 
  + [Paper]( https://doi.org/10.1093/bioinformatics/btv566)
* Minimap2, SAMtools - for read alignments to assembly, see details above

Installation:
We will also be using minimap2 and samtools which can be found in the ```polish``` Conda environment. Installation is outlined above.
Additionally, we will be running an offline BUSCO search using the appropriate database (eurotiales) for *A. fumigatus*. Details on downloading the appropriate BUSCO database can be found at: <https://busco.ezlab.org/>.
```
conda create -n asm-qc bioconda::busco=6.0.0 bioconda::sepp=4.5.5 bioconda::qualimap bioconda::seqkit
```


Note: if assembling an organism that is not part of the *Aspergilli* you may need to update the BUSCO database; if running on an alternative machine you will need to change to the appropriate download path. Additionally, read mapping is only for long-read sequence data, please see section on polishing to edit Minimap2 for short-read data.

Script for running assembly QC (derived from long-read data) can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm_qc.job
```


Usage:

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:20:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zAsmQC
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish/ont.polished.fasta
ONT=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/SRR28036069.trimmed.fastq
BUSCO_DB=~/../projects/fisher-local-software/live/busco_downloads
ASM_QC_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/asm_QC
#--------------------------------------------
# 3. Perform QC on polished assembly
#--------------------------------------------
mkdir $ASM_QC_DIR
conda activate asm-qc
# Traditional statistics
seqkit stats -a -N 50,90 $GENOME | sed 's/,//g' > $ASM_QC_DIR/basic_stats.txt
# Conserved universal orthologs
cd $TMPDIR
mkdir busco
busco -m genome -i $GENOME -o assembly --out_path busco -l eurotiales_odb10 \
--miniprot --offline \
--download_path ~/../projects/fisher-local-software/live/busco_downloads \
-f -c 10
cp ./busco/assembly/short_summary.specific.eurotiales_odb10.assembly.txt $ASM_QC_DIR/busco.txt
# Coverage 
## ONT read mapping
conda activate polish
minimap2 -a -x map-ont $GENOME $ONT | samtools sort -T tmp -o output.sorted.bam
samtools index output.sorted.bam
## Coverage calc.
conda activate asm-qc
mkdir $ASM_QC_DIR/qualimap
qualimap bamqc -nt 10 -bam output.sorted.bam -outdir $ASM_QC_DIR/qualimap \
--java-mem-size=80G
#--------------------------------------------
# 4. Summarise QC results
#--------------------------------------------
#num_seqs sum_len min_len avg_len max_len median GC% N50 N90
STATS=$(awk '{print $4,$5,$6,$7,$8,$10,$18,$20,$21'} $ASM_QC_DIR/basic_stats.txt \
| tail -n1 | sed 's/ /,/g')
#BUSCO Complete	| BUSCO Single |	BUSCO Duplicate	 | BUSCO Fragmented | 	BUSCO Missing	| BUSCO Num Genes
BUSCO=$(grep "C:" $ASM_QC_DIR/busco.txt | awk '{print $1}' | sed 's/\[/,/g' \
| sed 's/\]//g' | sed 's/[%|:|Aa-Zz]//g' | cut -d"," -f1-6)
COV=$(cat $ASM_QC_DIR/qualimap/genome_results.txt | grep "mean coverageData" \
| cut -d"=" -f2 | sed 's/ //g' | sed 's/X//g')
echo "${STATS},${COV},${BUSCO}" > $ASM_QC_DIR/overall_stats.csv
```
**Line 22** performs basic calculation of assembly statistics

**Lines 26-29** identifies the presence of single-copy conserved orthologs.
This provides an idea of genome completeness

**Lines 33-40** aligns reads to the genome (minimap2, samtools) and calculates
coverage (qualimap), providing an idea of read-depth

**Lines 45-52** extract formatted information from outputs and saves this as a 
single comma-separated line. Exact headers are described a combination of 
**Line 44 + coverage + Line 47**

# Basic gene prediction and annotation

Gene prediction and annotation are two major steps in understanding the constituent elements of a genome. Here we conjoin the two stages: prediction (identifying where genes are present) and annotation (characterizing identified genes). It's important to realize that these are two distinct steps. 

The pipeline below predominantly uses Funannotate and the tutorials from the respective wiki page. You will notice that many of the annotation steps we perform outside of Funannotate and combine at the end. As there are many steps in this script, please check the error/output files to troubleshoot problems as they arise.

Installation of components here was somewhat tricky, however I have tried to outline as much as I can for non-Fisher Lab (FL) users. Thus, I have separated out the installation segments for FL and non-FL users. It is important to note the database versions used, as they could impact your results.

Tools for prediction and annotation:

* Funannotate - fungal gene prediction
  + Version: 1.8.17
  + [GitHub](https://github.com/nextgenusfs/funannotate) 
  + [Wiki](http://funannotate.readthedocs.io/)
* InterProScan5 - annotation from various protein domain databases
  + Version: 5.67-99.0
  + [Wiki](https://interproscan-docs.readthedocs.io/en/v5/) 
  + [Paper](https://doi.org/10.1093/nar/gkaa977)
* EggNOG - functional annotation from orthologs
  + Version: 2.1.13
  + [GitHub](https://github.com/eggnogdb/eggnog-mapper) 
  + [Paper](https://doi.org/10.1093/molbev/msab293)
* AntiSMASH - secondary metabolite gene clusters
  + Version: 8.0.4
  + [GitHub](https://github.com/antismash/antismash) 
  + [Paper](https://doi.org/10.1093/nar/gkaf334)
* Infernal - RNA annotation
  + Version: 1.1.5
  + [Website](http://eddylab.org/infernal/) 
  + [Paper](https://doi.org/10.1093/bioinformatics/btt509)
* Seqkit, see details above

Installation (Non-/FL users):
```
conda create -n funannotate "python>=3.6,<3.9" bioconda::funannotate bioconda::eggnog-mapper
conda create -n antismash bioconda::antismash
conda create -n infernal bioconda::infernal
```
We have noticed unknown errors whilst running DIAMOND during the initial stages of funannotate.
If this appears to be happening to you, here is a copy of YAML file for the Conda environment.
Please save this below as ```funannotate.yml```:
**Note:** members of Fisher Lab can find a copy at ```~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/funannotate.yml```


<details>
<summary>Click to show Funannotate YAML</summary>

```
name: funannotate
channels:
  - conda-forge
  - bioconda
  - defaults
  - ursky
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=5.1=1_gnu
  - _r-mutex=1.0.1=anacondar_1
  - _x86_64-microarch-level=3=3_zen2
  - alsa-lib=1.2.15.2=hb03c661_0
  - antismash=7.1.0=pyh7e72e81_1
  - argcomplete=3.5.1=pyhd8ed1ab_0
  - aria2=1.37.0=h4e1c2bf_3
  - atk-1.0=2.38.0=hd4edc92_1
  - attr=2.5.2=h39aace5_0
  - attrs=24.2.0=pyh71513ae_0
  - augustus=3.5.0=pl5321h9716f88_9
  - bamtools=2.5.3=he132191_0
  - bcbio-gff=0.7.1=pyhdfd78af_3
  - bedtools=2.31.1=h13024bc_3
  - binutils_impl_linux-64=2.45=default_hfdba357_105
  - binutils_linux-64=2.45=default_h4852527_105
  - bioconductor-annotate=1.80.0=r43hdfd78af_0
  - bioconductor-annotationdbi=1.64.1=r43hdfd78af_0
  - bioconductor-apeglm=1.24.0=r43hf17093f_1
  - bioconductor-biobase=2.62.0=r43ha9d7317_3
  - bioconductor-biocfilecache=2.10.1=r43hdfd78af_0
  - bioconductor-biocgenerics=0.48.1=r43hdfd78af_2
  - bioconductor-biocio=1.12.0=r43hdfd78af_0
  - bioconductor-biocparallel=1.36.0=r43hf17093f_2
  - bioconductor-biomart=2.58.0=r43hdfd78af_0
  - bioconductor-biostrings=2.70.1=r43ha9d7317_2
  - bioconductor-ctc=1.76.0=r43hdfd78af_0
  - bioconductor-data-packages=20250625=hdfd78af_0
  - bioconductor-delayedarray=0.28.0=r43ha9d7317_2
  - bioconductor-deseq2=1.42.0=r43hf17093f_2
  - bioconductor-dexseq=1.48.0=r43hdfd78af_2
  - bioconductor-edger=4.0.16=r43hf17093f_1
  - bioconductor-genefilter=1.84.0=r43ha1e849b_1
  - bioconductor-genelendatabase=1.38.0=r43hdfd78af_0
  - bioconductor-geneplotter=1.80.0=r43hdfd78af_0
  - bioconductor-genomeinfodb=1.38.1=r43hdfd78af_1
  - bioconductor-genomeinfodbdata=1.2.11=r43hdfd78af_1
  - bioconductor-genomicalignments=1.38.0=r43ha9d7317_1
  - bioconductor-genomicfeatures=1.54.1=r43hdfd78af_0
  - bioconductor-genomicranges=1.54.1=r43ha9d7317_2
  - bioconductor-go.db=3.18.0=r43hdfd78af_0
  - bioconductor-goseq=1.54.0=r43hdfd78af_0
  - bioconductor-iranges=2.36.0=r43ha9d7317_2
  - bioconductor-keggrest=1.42.0=r43hdfd78af_0
  - bioconductor-limma=3.58.1=r43ha9d7317_1
  - bioconductor-matrixgenerics=1.14.0=r43hdfd78af_3
  - bioconductor-qvalue=2.34.0=r43hdfd78af_0
  - bioconductor-rhtslib=2.4.0=r43ha9d7317_2
  - bioconductor-rsamtools=2.18.0=r43hf17093f_2
  - bioconductor-rtracklayer=1.62.0=r43ha9d7317_1
  - bioconductor-s4arrays=1.2.0=r43ha9d7317_2
  - bioconductor-s4vectors=0.40.2=r43ha9d7317_2
  - bioconductor-seqlogo=1.68.0=r43hdfd78af_0
  - bioconductor-sparsearray=1.2.2=r43ha9d7317_2
  - bioconductor-summarizedexperiment=1.32.0=r43hdfd78af_0
  - bioconductor-xvector=0.42.0=r43ha9d7317_2
  - bioconductor-zlibbioc=1.48.0=r43ha9d7317_2
  - biopython=1.79=py38h0a891b7_3
  - blast=2.17.0=h66d330f_0
  - boost-cpp=1.85.0=h44aadfe_2
  - bowtie2=2.5.4=he96a11b_7
  - brawn=1.0.2=pyhdfd78af_0
  - brotli=1.1.0=hb03c661_4
  - brotli-bin=1.1.0=hb03c661_4
  - brotli-python=1.0.9=py38hfa26641_8
  - bwidget=1.10.1=ha770c72_1
  - bx-python=0.13.0=py38h737be40_0
  - bzip2=1.0.8=hda65f42_8
  - c-ares=1.34.6=hb03c661_0
  - ca-certificates=2026.1.4=hbd8a1cb_0
  - cairo=1.18.0=h3faef2a_0
  - cdbtools=0.99=h077b44d_12
  - certifi=2024.8.30=pyhd8ed1ab_0
  - cffi=1.17.1=py38h1fdaa30_0
  - charset-normalizer=3.4.0=pyhd8ed1ab_0
  - codingquarry=2.0=py38h2123bcc_9
  - contourpy=1.1.1=py38h7f3f72f_1
  - coreutils=9.8=hee96239_0
  - curl=8.18.0=h4e3cde8_0
  - cycler=0.12.1=pyhd8ed1ab_0
  - cyrus-sasl=2.1.28=hd9c7081_0
  - dbus=1.13.18=hb2f20db_0
  - diamond=2.1.18=h13889ed_0
  - distro=1.9.0=pyhd8ed1ab_0
  - docopt=0.6.2=py_1
  - easel=0.49=h7b50bb2_2
  - eggnog-mapper=2.1.13=pyhdfd78af_1
  - entrez-direct=24.0=he881be0_0
  - et_xmlfile=2.0.0=pyhd8ed1ab_0
  - ete3=3.1.3=pyhd8ed1ab_0
  - evidencemodeler=1.1.1=0
  - exonerate=2.4.0=hf34a1b8_7
  - expat=2.7.3=hecca717_0
  - fasta3=36.3.8=hec16e2b_7
  - fasttree=2.1.11=h7b50bb2_5
  - font-ttf-dejavu-sans-mono=2.37=hab24e00_0
  - font-ttf-inconsolata=3.000=h77eed37_0
  - font-ttf-source-code-pro=2.038=h77eed37_0
  - font-ttf-ubuntu=0.83=h77eed37_3
  - fontconfig=2.15.0=h7e30c49_1
  - fonts-conda-ecosystem=1=0
  - fonts-conda-forge=1=hc364b38_1
  - fonttools=4.53.1=py38h2019614_0
  - freetype=2.14.1=ha770c72_0
  - fribidi=1.0.16=hb03c661_0
  - ftpretty=0.4.0=pyhd8ed1ab_0
  - funannotate=1.8.17=pyhdfd78af_5
  - gawk=5.3.1=hcd3d067_0
  - gcc_impl_linux-64=13.4.0=hf787b08_16
  - gcc_linux-64=13.4.0=h0a5b801_17
  - gdbm=1.18=h0a1914f_2
  - gdk-pixbuf=2.42.10=h829c605_4
  - gettext=0.25.1=h3f43e3d_1
  - gettext-tools=0.25.1=h3f43e3d_1
  - gfortran_impl_linux-64=13.4.0=hb485fb8_16
  - giflib=5.2.2=hd590300_0
  - glib=2.78.1=hfc55251_0
  - glib-tools=2.78.1=hfc55251_0
  - glimmerhmm=3.0.4=pl5321h503566f_10
  - glpk=5.0=h445213a_0
  - gmap=2025.07.31=pl5321hb1d24b7_1
  - gmp=6.3.0=hac33072_2
  - gnutls=3.7.8=hf3e180e_0
  - goatools=1.4.12=pyhd8ed1ab_0
  - graphite2=1.3.14=hecca717_2
  - graphviz=9.0.0=h78e8752_1
  - gsl=2.8=hbf7d49c_1
  - gst-plugins-base=1.22.7=h8e1006c_0
  - gstreamer=1.22.7=h98fc4e7_0
  - gtk2=2.24.33=h90689f9_2
  - gts=0.7.6=h977cf35_4
  - gxx_impl_linux-64=13.4.0=h6a38259_16
  - h2=4.1.0=py38h578d9bd_0
  - harfbuzz=8.3.0=h3d44ed6_0
  - hdf5=1.14.3=nompi_h2d575fe_109
  - helperlibs=0.2.1=py_0
  - hisat2=2.2.1=h503566f_8
  - hmmer=3.4=h503566f_3
  - hmmer2=2.3.2=h7b50bb2_11
  - hpack=4.0.0=pyh9f0ad1d_0
  - htseq=2.0.5=py38h38ace90_2
  - htslib=1.22.1=h566b1c6_0
  - hyperframe=6.0.1=pyhd8ed1ab_0
  - icu=73.2=h59595ed_0
  - idna=3.10=pyhd8ed1ab_0
  - importlib-resources=6.4.5=pyhd8ed1ab_0
  - importlib_resources=6.4.5=pyhd8ed1ab_0
  - infernal=1.1.5=pl5321h7b50bb2_4
  - jansson=2.14.1=hb03c661_2
  - jinja2=3.1.4=pyhd8ed1ab_0
  - joblib=1.4.2=pyhd8ed1ab_0
  - jq=1.8.1=h73b1eb8_0
  - jsoncpp=1.9.6=hf42df4d_1
  - jsonschema=4.23.0=pyhd8ed1ab_0
  - jsonschema-specifications=2024.10.1=pyhd8ed1ab_0
  - k8=1.2=he8db53b_6
  - kallisto=0.46.1=h6f0a7f7_1
  - kernel-headers_linux-64=4.18.0=he073ed8_9
  - keyutils=1.6.3=hb9d3cd8_0
  - kiwisolver=1.4.5=py38h7f3f72f_1
  - kmer-jellyfish=2.3.1=py38h2123bcc_4
  - krb5=1.21.3=h659f571_0
  - lame=3.100=h166bdaf_1003
  - lcms2=2.18=h0c24ade_0
  - ld_impl_linux-64=2.45=default_hbd61a6d_105
  - lerc=4.0.0=h0aef613_1
  - libabseil=20250127.0=cxx17_hbbce691_0
  - libaec=1.1.4=h3f801dc_0
  - libamd=3.3.3=haaf9dc3_7100102
  - libasprintf=0.25.1=h3f43e3d_1
  - libasprintf-devel=0.25.1=h3f43e3d_1
  - libblas=3.9.0=20_linux64_openblas
  - libboost=1.85.0=hba137d9_2
  - libboost-devel=1.85.0=h00ab1b0_2
  - libboost-headers=1.85.0=ha770c72_2
  - libbrotlicommon=1.1.0=hb03c661_4
  - libbrotlidec=1.1.0=hb03c661_4
  - libbrotlienc=1.1.0=hb03c661_4
  - libbtf=2.3.2=h32481e8_7100102
  - libcamd=3.3.3=h32481e8_7100102
  - libcap=2.77=h3ff7636_0
  - libcblas=3.9.0=20_linux64_openblas
  - libccolamd=3.3.4=h32481e8_7100102
  - libcholmod=5.3.1=h59ddab4_7100102
  - libclang=15.0.7=default_h127d8a8_5
  - libclang13=15.0.7=default_h5d6823c_5
  - libcolamd=3.3.4=h32481e8_7100102
  - libcups=2.3.3=hb8b1518_5
  - libcurl=8.18.0=h4e3cde8_0
  - libcxsparse=4.4.1=h32481e8_7100102
  - libdb=6.2.32=h9c3ff4c_0
  - libdeflate=1.25=h17f619e_0
  - libedit=3.1.20250104=pl5321h7949ede_0
  - libev=4.33=hd590300_2
  - libevent=2.1.12=hf998b51_1
  - libexpat=2.7.3=hecca717_0
  - libffi=3.4.6=h2dba641_1
  - libflac=1.5.0=he200343_1
  - libfreetype=2.14.1=ha770c72_0
  - libfreetype6=2.14.1=h73754d4_0
  - libgcc=15.2.0=he0feb66_16
  - libgcc-devel_linux-64=13.4.0=hd1d28cc_116
  - libgcc-ng=15.2.0=h69a702a_16
  - libgd=2.3.3=h119a65a_9
  - libgettextpo=0.25.1=h3f43e3d_1
  - libgettextpo-devel=0.25.1=h3f43e3d_1
  - libgfortran=15.2.0=h69a702a_16
  - libgfortran-ng=15.2.0=h69a702a_16
  - libgfortran5=15.2.0=h68bc16d_16
  - libglib=2.78.1=hebfc3b9_0
  - libgomp=15.2.0=he0feb66_16
  - libhwloc=2.12.1=default_h3d81e11_1000
  - libiconv=1.18=h3b78370_2
  - libidn11=1.34=h1cef754_0
  - libidn2=2.3.8=hfac485b_1
  - libjemalloc=5.3.0=h5888daf_1
  - libjpeg-turbo=3.1.3=h47b2149_0
  - libklu=2.3.5=hf24d653_7100102
  - liblapack=3.9.0=20_linux64_openblas
  - liblapacke=3.9.0=20_linux64_openblas
  - libldl=3.3.2=h32481e8_7100102
  - libllvm15=15.0.7=ha7bfdaf_5
  - liblzma=5.8.1=hb9d3cd8_2
  - liblzma-devel=5.8.1=hb9d3cd8_2
  - libnghttp2=1.67.1=h697f920_0
  - libnsl=2.0.1=hb9d3cd8_1
  - libntlm=1.8=hb9d3cd8_0
  - libogg=1.3.5=hd0c01bc_1
  - libopenblas=0.3.25=pthreads_h413a1c8_0
  - libopenssl-static=3.6.0=hb03c661_0
  - libopus=1.5.2=hd0c01bc_0
  - libparu=1.0.0=h17147ab_7100102
  - libpng=1.6.53=h421ea60_0
  - libpq=16.11=hc183893_2
  - libprotobuf=5.29.3=h501fc15_0
  - librbio=4.3.4=h32481e8_7100102
  - librsvg=2.56.3=he3f83f7_1
  - libsanitizer=13.4.0=h2a15e64_16
  - libsass=0.22.0=py38h17151c0_1
  - libsndfile=1.2.2=hc7d488a_2
  - libspex=3.2.3=had10066_7100102
  - libspqr=4.3.4=h852d39f_7100102
  - libsqlite=3.51.2=h0c1763c_0
  - libssh2=1.11.1=hcf80075_0
  - libstdcxx=15.2.0=h934c35e_16
  - libstdcxx-devel_linux-64=13.4.0=h6963c3b_116
  - libstdcxx-ng=15.2.0=hdf11a46_16
  - libsuitesparseconfig=7.10.1=h92d6892_7100102
  - libsystemd0=257.10=hd0affe5_3
  - libtasn1=4.21.0=h27ab2c4_0
  - libtiff=4.7.1=h9d88235_1
  - libumfpack=6.3.5=heb53515_7100102
  - libunistring=0.9.10=h7f98852_0
  - libuuid=2.41.3=h5347b49_0
  - libuv=1.51.0=hb03c661_1
  - libvorbis=1.3.7=h54a6638_2
  - libwebp=1.6.0=h9635ea4_0
  - libwebp-base=1.6.0=hd42ef1d_0
  - libxcb=1.15=h0b41bf4_0
  - libxcrypt=4.4.36=hd590300_1
  - libxkbcommon=1.7.0=h662e7e4_0
  - libxml2=2.13.9=h2c43086_0
  - libxslt=1.1.43=h7a3aeb2_0
  - libzlib=1.3.1=hb9d3cd8_2
  - lighttpd=1.4.72=lua54h2766e62_2
  - lp_solve=5.5.2.11=hd590300_0
  - lua=5.4.8=h03e1676_1
  - lxml=5.3.0=py38hd8ce619_0
  - lz4-c=1.9.4=hcb278e6_0
  - mafft=7.526=h4bc722e_0
  - make=4.4.1=hb9d3cd8_2
  - mariadb-connector-c=3.4.8=h6dee460_0
  - markdown-it-py=3.0.0=pyhd8ed1ab_0
  - markupsafe=2.1.5=py38h01eb140_0
  - matplotlib-base=3.7.3=py38h58ed7fa_0
  - mdurl=0.1.2=pyhd8ed1ab_0
  - meme=5.5.9=pl5321h1ca524f_0
  - metis=5.1.0=hd0bcaf9_1007
  - minimap2=2.30=h577a1d6_0
  - mmseqs2=18.8cc5c=hd6d6fdc_0
  - moods=1.9.4.2=py38h2123bcc_2
  - mpfr=4.2.1=h90cbb55_3
  - mpg123=1.32.9=hc50e24c_0
  - mpi=1.0=openmpi
  - munkres=1.1.4=pyh9f0ad1d_0
  - mysql=8.4.0=h721767e_2
  - mysql-common=8.0.33=hf1915f5_6
  - mysql-connector-c=6.1.11=h659d440_1008
  - mysql-libs=8.0.33=hca2cd23_6
  - natsort=8.4.0=pyhd8ed1ab_0
  - ncbi-vdb=3.3.0=h9948957_0
  - ncurses=6.5=h2d0b736_3
  - nettle=3.8.1=hc379101_1
  - nodejs=22.6.0=h6d9b948_0
  - nrpys=0.1.1=pyha8f3691_3
  - nspr=4.38=h29cc59b_0
  - nss=3.118=h445c969_0
  - numpy=1.24.4=py38h59b608b_0
  - oniguruma=6.9.10=hb9d3cd8_0
  - openjdk=21.0.2=haa376d0_0
  - openjpeg=2.5.4=h55fea9a_0
  - openldap=2.6.10=he970967_0
  - openmpi=4.1.6=hc5af2df_101
  - openpyxl=3.1.5=py38h01eb140_0
  - openssl=3.6.0=h26f9b46_0
  - p11-kit=0.24.1=hc5aa10d_0
  - packaging=25.0=pyh29332c3_1
  - pandas=2.0.3=py38h01efb38_1
  - pandoc=3.8.3=ha770c72_0
  - pango=1.50.14=ha41ecd1_2
  - pasa=2.5.3=h9948957_2
  - patsy=0.5.6=pyhd8ed1ab_0
  - pblat=2.5.1=h41f7678_4
  - pcre=8.45=h9c3ff4c_0
  - pcre2=10.40=hc3806b6_0
  - perl=5.32.1=7_hd590300_perl5
  - perl-app-cpanminus=1.7048=pl5321hd8ed1ab_0
  - perl-archive-tar=3.04=pl5321hdfd78af_0
  - perl-b-cow=0.007=pl5321hb9d3cd8_1
  - perl-base=2.23=pl5321hd8ed1ab_0
  - perl-business-isbn=3.007=pl5321hd8ed1ab_0
  - perl-business-isbn-data=20210112.006=pl5321hd8ed1ab_0
  - perl-carp=1.50=pl5321hd8ed1ab_0
  - perl-cgi=4.56=pl5321h7b50bb2_2
  - perl-class-data-inheritable=0.09=pl5321ha770c72_0
  - perl-class-inspector=1.36=pl5321hdfd78af_0
  - perl-class-method-modifiers=2.13=pl5321ha770c72_0
  - perl-clone=0.46=pl5321hb9d3cd8_1
  - perl-clone-choose=0.010=pl5321hd8ed1ab_0
  - perl-common-sense=3.75=pl5321hd8ed1ab_0
  - perl-compress-raw-bzip2=2.214=pl5321hda65f42_0
  - perl-compress-raw-zlib=2.214=pl5321h4dac143_0
  - perl-config-general=2.67=pl5321hdfd78af_0
  - perl-constant=1.33=pl5321hd8ed1ab_0
  - perl-convert-binhex=1.125=pl5321hdfd78af_2
  - perl-data-dump=1.25=pl5321h7b50bb2_2
  - perl-data-dumper=2.183=pl5321hb9d3cd8_1
  - perl-db_file=1.858=pl5321hb9d3cd8_0
  - perl-dbd-mysql=5.013=pl5321h0a44790_0
  - perl-dbd-sqlite=1.76=pl5321h7b50bb2_2
  - perl-dbi=1.647=pl5321hb03c661_0
  - perl-devel-checklib=1.16=pl5321h7b50bb2_1
  - perl-devel-stacktrace=2.04=pl5321h296ab09_0
  - perl-digest-hmac=1.05=pl5321hdfd78af_0
  - perl-digest-md5=2.59=pl5321hb9d3cd8_3
  - perl-email-date-format=1.006=pl5321hd8ed1ab_0
  - perl-encode=3.21=pl5321hb9d3cd8_1
  - perl-encode-locale=1.05=pl5321hdfd78af_7
  - perl-exception-class=1.45=pl5321ha770c72_0
  - perl-exporter=5.74=pl5321hd8ed1ab_0
  - perl-exporter-tiny=1.002002=pl5321hd8ed1ab_0
  - perl-extutils-makemaker=7.70=pl5321hd8ed1ab_0
  - perl-file-listing=6.16=pl5321hdfd78af_0
  - perl-file-path=2.18=pl5321hd8ed1ab_0
  - perl-file-spec=3.48_01=pl5321hdfd78af_2
  - perl-file-temp=0.2304=pl5321hd8ed1ab_0
  - perl-file-util=4.201720=pl5321hdfd78af_0
  - perl-file-which=1.24=pl5321hd8ed1ab_0
  - perl-getopt-long=2.58=pl5321hdfd78af_0
  - perl-hash-merge=0.302=pl5321hd8ed1ab_0
  - perl-html-parser=3.81=pl5321h4ac6f70_1
  - perl-html-tagset=3.24=pl5321hdfd78af_0
  - perl-html-template=2.97=pl5321hdfd78af_2
  - perl-html-tree=5.07=pl5321hdfd78af_2
  - perl-http-cookiejar-lwp=0.014=pl5321hdfd78af_0
  - perl-http-cookies=6.11=pl5321hdfd78af_0
  - perl-http-daemon=6.16=pl5321hdfd78af_0
  - perl-http-date=6.06=pl5321hdfd78af_0
  - perl-http-message=7.01=pl5321hdfd78af_0
  - perl-http-negotiate=6.01=pl5321hdfd78af_4
  - perl-inc-latest=0.500=pl5321ha770c72_0
  - perl-io-compress=2.214=pl5321h503566f_0
  - perl-io-html=1.004=pl5321hdfd78af_0
  - perl-io-sessiondata=1.03=pl5321hdfd78af_2
  - perl-io-socket-ssl=2.075=pl5321hd8ed1ab_0
  - perl-io-zlib=1.15=pl5321hdfd78af_1
  - perl-json=4.10=pl5321hdfd78af_1
  - perl-json-xs=4.04=pl5321h9948957_0
  - perl-lib=0.63=pl5321hd8ed1ab_0
  - perl-libwww-perl=6.81=pl5321hdfd78af_0
  - perl-list-moreutils=0.430=pl5321hdfd78af_0
  - perl-list-moreutils-xs=0.430=pl5321h7b50bb2_5
  - perl-local-lib=2.000029=pl5321hdfd78af_0
  - perl-log-log4perl=1.56=pl5321hd8ed1ab_0
  - perl-logger-simple=2.0=pl5321hdfd78af_1
  - perl-lwp-mediatypes=6.04=pl5321hdfd78af_1
  - perl-lwp-protocol-https=6.14=pl5321hdfd78af_1
  - perl-mailtools=2.22=pl5321hdfd78af_0
  - perl-math-cdf=0.1=pl5321h7b50bb2_11
  - perl-math-utils=1.14=pl5321hdfd78af_0
  - perl-mce=1.902=pl5321hdfd78af_0
  - perl-mime-base64=3.16=pl5321hb9d3cd8_3
  - perl-mime-lite=3.030=pl5321hdfd78af_2
  - perl-mime-tools=5.515=pl5321hdfd78af_0
  - perl-mime-types=2.30=pl5321hdfd78af_0
  - perl-module-build=0.4234=pl5321ha770c72_1
  - perl-module-load=0.34=pl5321hdfd78af_0
  - perl-moo=2.005004=pl5321ha770c72_0
  - perl-mozilla-ca=20250602=pl5321hdfd78af_0
  - perl-net-http=6.24=pl5321hdfd78af_0
  - perl-net-ssleay=1.94=pl5321hf672d98_1
  - perl-ntlm=1.09=pl5321hdfd78af_5
  - perl-object-insideout=4.05=pl5321hdfd78af_1
  - perl-parallel-forkmanager=2.04=pl5321hdfd78af_0
  - perl-parent=0.243=pl5321hd8ed1ab_0
  - perl-pathtools=3.75=pl5321hb9d3cd8_2
  - perl-role-tiny=2.002004=pl5321ha770c72_0
  - perl-scalar-list-utils=1.70=pl5321hb03c661_0
  - perl-scalar-util-numeric=0.40=pl5321h7b50bb2_7
  - perl-soap-lite=1.27=pl5321hdfd78af_0
  - perl-socket=2.027=pl5321h5c03b87_6
  - perl-storable=3.15=pl5321hb9d3cd8_2
  - perl-sub-quote=2.006006=pl5321ha770c72_0
  - perl-sys-info=0.7811=pl5321hdfd78af_1
  - perl-sys-info-base=0.7807=pl5321hdfd78af_1
  - perl-sys-info-driver-linux=0.7911=pl5321hdfd78af_0
  - perl-task-weaken=1.06=pl5321hdfd78af_1
  - perl-test-fatal=0.016=pl5321hdfd78af_0
  - perl-test-harness=3.44=pl5321hd8ed1ab_0
  - perl-test-needs=0.002009=pl5321hd8ed1ab_0
  - perl-test-nowarnings=1.06=pl5321ha770c72_0
  - perl-test-pod=1.52=pl5321hdfd78af_1
  - perl-test-requiresinternet=0.05=pl5321hdfd78af_1
  - perl-test-warnings=0.031=pl5321ha770c72_0
  - perl-text-soundex=3.05=pl5321hd8ed1ab_1001
  - perl-text-template-simple=0.91=pl5321hdfd78af_1
  - perl-time-hires=1.9764=pl5321h7b50bb2_6
  - perl-time-local=1.35=pl5321hdfd78af_0
  - perl-timedate=2.33=pl5321hdfd78af_2
  - perl-try-tiny=0.32=pl5321ha770c72_1
  - perl-types-serialiser=1.01=pl5321hdfd78af_0
  - perl-unix-processors=2.046=pl5321hb9d3cd8_1002
  - perl-uri=5.34=pl5321ha770c72_0
  - perl-url-encode=0.03=pl5321h9ee0642_1
  - perl-www-robotrules=6.02=pl5321hdfd78af_4
  - perl-xml-namespacesupport=1.12=pl5321hd8ed1ab_0
  - perl-xml-parser=2.44_01=pl5321hf2e2c51_1004
  - perl-xml-sax=1.02=pl5321hd8ed1ab_0
  - perl-xml-sax-base=1.09=pl5321hd8ed1ab_0
  - perl-xml-sax-expat=0.51=pl5321hd8ed1ab_0
  - perl-xml-simple=2.25=pl5321hdfd78af_2
  - perl-yaml=1.30=pl5321hdfd78af_0
  - phyml=3.3.20220408=h9bc3f66_3
  - pigz=2.8=h421ea60_2
  - pillow=10.3.0=py38h9e66945_0
  - pip=24.3.1=pyh8b19718_0
  - pixman=0.46.4=h54a6638_1
  - pkgutil-resolve-name=1.3.10=pyhd8ed1ab_1
  - platformdirs=4.3.6=pyhd8ed1ab_0
  - ply=3.11=pyhd8ed1ab_2
  - pooch=1.8.2=pyhd8ed1ab_0
  - prodigal=2.6.3=h577a1d6_11
  - proteinortho=6.3.6=h2b77389_0
  - psutil=6.0.0=py38hfb59056_0
  - pthread-stubs=0.4=hb9d3cd8_1002
  - pulseaudio-client=16.1=hb77b528_5
  - pycparser=2.22=pyhd8ed1ab_0
  - pydot=3.0.1=py38h578d9bd_0
  - pygments=2.18.0=pyhd8ed1ab_0
  - pyparsing=3.1.4=pyhd8ed1ab_0
  - pyqt=5.15.10=py38h6a678d5_0
  - pyqt5-sip=12.13.0=py38h5eee18b_0
  - pysam=0.22.1=py38h2d19e03_2
  - pysocks=1.7.1=py38h578d9bd_5
  - pysvg-py3=0.2.2.post3=py38h578d9bd_7
  - python=3.8.20=h4a871b0_2_cpython
  - python-dateutil=2.9.0post0=py38h06a4308_2
  - python-tzdata=2025.2=pyhd3eb1b0_0
  - python_abi=3.8=6_cp38
  - pytz=2024.2=pyhd8ed1ab_0
  - pyyaml=6.0.2=py38h2019614_0
  - qt-main=5.15.8=h82b777d_17
  - r-abind=1.4_5=r43hc72bb7e_1006
  - r-amap=0.8_20=r43he59cc3e_0
  - r-ape=5.8_1=r43h3704496_1
  - r-argparse=2.2.5=r43hc72bb7e_0
  - r-ashr=2.2_63=r43h93ab643_2
  - r-askpass=1.2.1=r43h2b5f3a1_0
  - r-assertthat=0.2.1=r43hc72bb7e_5
  - r-backports=1.5.0=r43hb1dbf0f_1
  - r-base=4.3.2=h93585b2_0
  - r-base64enc=0.1_3=r43hb1dbf0f_1007
  - r-bbmle=1.0.25.1=r43hc72bb7e_1
  - r-bdsmatrix=1.3_7=r43h2b5f3a1_2
  - r-bh=1.87.0_1=r43hc72bb7e_0
  - r-biasedurn=2.0.12=r43h0d4f4ea_1
  - r-bit=4.6.0=r43h2b5f3a1_0
  - r-bit64=4.6.0_1=r43h2b5f3a1_0
  - r-bitops=1.0_9=r43h2b5f3a1_0
  - r-blob=1.2.4=r43hc72bb7e_2
  - r-broom=1.0.9=r43hc72bb7e_0
  - r-bslib=0.9.0=r43hc72bb7e_0
  - r-cachem=1.1.0=r43hb1dbf0f_1
  - r-callr=3.7.6=r43hc72bb7e_1
  - r-catools=1.18.3=r43h93ab643_0
  - r-cellranger=1.1.0=r43hc72bb7e_1007
  - r-cli=3.6.5=r43h93ab643_0
  - r-clipr=0.8.0=r43hc72bb7e_3
  - r-cluster=2.1.8.1=r43hb67ce94_0
  - r-coda=0.19_4.1=r43hc72bb7e_1
  - r-codetools=0.2_20=r43hc72bb7e_1
  - r-colorspace=2.1_1=r43hdb488b9_0
  - r-conflicted=1.2.0=r43h785f33e_2
  - r-cpp11=0.5.2=r43h785f33e_1
  - r-crayon=1.5.3=r43hc72bb7e_1
  - r-curl=7.0.0=r43h10955f1_0
  - r-data.table=1.17.8=r43h1c8cec4_0
  - r-dbi=1.2.3=r43hc72bb7e_1
  - r-dbplyr=2.5.1=r43hc72bb7e_0
  - r-digest=0.6.37=r43h0d4f4ea_0
  - r-dplyr=1.1.4=r43h0d4f4ea_1
  - r-dtplyr=1.3.2=r43hc72bb7e_0
  - r-ellipsis=0.3.2=r43hb1dbf0f_3
  - r-emdbook=1.3.14=r43hc72bb7e_0
  - r-etrunct=0.1=r43hc72bb7e_1006
  - r-evaluate=1.0.5=r43hc72bb7e_0
  - r-fansi=1.0.6=r43hb1dbf0f_1
  - r-farver=2.1.2=r43ha18555a_1
  - r-fastcluster=1.3.0=r43h5888daf_0
  - r-fastmap=1.2.0=r43ha18555a_1
  - r-fastmatch=1.1_6=r43h2b5f3a1_0
  - r-filelock=1.0.3=r43hb1dbf0f_1
  - r-findpython=1.0.9=r43hc72bb7e_0
  - r-fontawesome=0.5.3=r43hc72bb7e_0
  - r-forcats=1.0.0=r43hc72bb7e_2
  - r-formatr=1.14=r43hc72bb7e_2
  - r-fs=1.6.6=r43h93ab643_0
  - r-futile.logger=1.4.3=r43hc72bb7e_1006
  - r-futile.options=1.0.1=r43hc72bb7e_1005
  - r-gargle=1.6.0=r43h785f33e_0
  - r-generics=0.1.4=r43hc72bb7e_0
  - r-ggplot2=3.5.2=r43hc72bb7e_0
  - r-glue=1.8.0=r43h2b5f3a1_0
  - r-googledrive=2.1.2=r43hc72bb7e_0
  - r-googlesheets4=1.1.2=r43h785f33e_0
  - r-gplots=3.2.0=r43hc72bb7e_0
  - r-gtable=0.3.6=r43hc72bb7e_0
  - r-gtools=3.9.5=r43hb1dbf0f_1
  - r-haven=2.5.5=r43h6d565e7_0
  - r-highr=0.11=r43hc72bb7e_1
  - r-hms=1.1.3=r43hc72bb7e_2
  - r-htmltools=0.5.8.1=r43ha18555a_1
  - r-httr=1.4.7=r43hc72bb7e_1
  - r-hwriter=1.3.2.1=r43hc72bb7e_3
  - r-ids=1.0.1=r43hc72bb7e_4
  - r-igraph=2.1.4=r43hadbbdbc_0
  - r-invgamma=1.2=r43hc72bb7e_0
  - r-irlba=2.3.5.1=r43h0d28552_3
  - r-isoband=0.2.7=r43ha18555a_3
  - r-jquerylib=0.1.4=r43hc72bb7e_3
  - r-jsonlite=2.0.0=r43h2b5f3a1_0
  - r-kernsmooth=2.23_26=r43h8461fee_0
  - r-knitr=1.50=r43hc72bb7e_0
  - r-labeling=0.4.3=r43hc72bb7e_1
  - r-lambda.r=1.2.4=r43hc72bb7e_4
  - r-lattice=0.22_7=r43h2b5f3a1_0
  - r-lifecycle=1.0.4=r43hc72bb7e_1
  - r-locfit=1.5_9.12=r43h2b5f3a1_0
  - r-lubridate=1.9.4=r43h2b5f3a1_0
  - r-magrittr=2.0.3=r43hb1dbf0f_3
  - r-mass=7.3_60.0.1=r43hb1dbf0f_1
  - r-matrix=1.6_5=r43he966344_1
  - r-matrixstats=1.5.0=r43h2b5f3a1_0
  - r-memoise=2.0.1=r43hc72bb7e_3
  - r-mgcv=1.9_3=r43h2ae2be5_0
  - r-mime=0.13=r43h2b5f3a1_0
  - r-mixsqp=0.3_54=r43hb79369c_3
  - r-modelr=0.1.11=r43hc72bb7e_2
  - r-munsell=0.5.1=r43hc72bb7e_1
  - r-mvtnorm=1.3_3=r43h9ad1c49_0
  - r-nlme=3.1_168=r43hb67ce94_0
  - r-numderiv=2016.8_1.1=r43hc72bb7e_6
  - r-openssl=2.3.3=r43he8289e2_0
  - r-phangorn=2.12.1=r43h9124c8a_2
  - r-pillar=1.11.0=r43hc72bb7e_0
  - r-pkgconfig=2.0.3=r43hc72bb7e_4
  - r-plogr=0.2.0=r43hc72bb7e_1006
  - r-plyr=1.8.9=r43h3697838_2
  - r-png=0.1_8=r43h21f035c_2
  - r-prettyunits=1.2.0=r43hc72bb7e_1
  - r-processx=3.8.6=r43h2b5f3a1_0
  - r-progress=1.2.3=r43hc72bb7e_1
  - r-ps=1.9.1=r43h2b5f3a1_0
  - r-purrr=1.1.0=r43h54b55ab_0
  - r-quadprog=1.5_8=r43hc2011d3_6
  - r-r6=2.6.1=r43hc72bb7e_0
  - r-ragg=1.5.0=r43h9f1dc4d_0
  - r-rappdirs=0.3.3=r43hb1dbf0f_3
  - r-rcolorbrewer=1.1_3=r43h785f33e_3
  - r-rcpp=1.1.0=r43h93ab643_0
  - r-rcpparmadillo=15.0.2_1=r43h3704496_0
  - r-rcppeigen=0.3.4.0.2=r43hb79369c_0
  - r-rcppnumerical=0.6_0=r43h0d4f4ea_1
  - r-rcurl=1.98_1.17=r43hb79926e_0
  - r-readr=2.1.5=r43h0d4f4ea_1
  - r-readxl=1.4.5=r43h328fee5_0
  - r-rematch=2.0.0=r43hc72bb7e_1
  - r-rematch2=2.1.2=r43hc72bb7e_4
  - r-reprex=2.1.1=r43hc72bb7e_1
  - r-reshape2=1.4.4=r43h3697838_5
  - r-restfulr=0.0.15=r43h56115f1_4
  - r-rjson=0.2.23=r43h93ab643_0
  - r-rlang=1.1.6=r43h93ab643_0
  - r-rmarkdown=2.29=r43hc72bb7e_0
  - r-rsqlite=2.4.3=r43h3697838_0
  - r-rstudioapi=0.17.1=r43hc72bb7e_0
  - r-rvest=1.0.5=r43hc72bb7e_0
  - r-sass=0.4.10=r43h93ab643_0
  - r-scales=1.4.0=r43hc72bb7e_0
  - r-selectr=0.4_2=r43hc72bb7e_4
  - r-sm=2.2_6.0=r43hc4980d5_1
  - r-snow=0.4_4=r43hc72bb7e_3
  - r-squarem=2021.1=r43hc72bb7e_3
  - r-statmod=1.5.0=r43ha36c22a_2
  - r-stringi=1.8.4=r43ha8ce623_2
  - r-stringr=1.5.2=r43h785f33e_0
  - r-survival=3.8_3=r43h2b5f3a1_0
  - r-sys=3.4.3=r43h2b5f3a1_0
  - r-systemfonts=1.2.3=r43h74f4acd_0
  - r-textshaping=0.3.7=r43hd87b9d6_0
  - r-tibble=3.3.0=r43h2b5f3a1_0
  - r-tidyr=1.3.1=r43h0d4f4ea_1
  - r-tidyselect=1.2.1=r43hc72bb7e_1
  - r-tidyverse=2.0.0=r43h785f33e_2
  - r-timechange=0.3.0=r43ha18555a_1
  - r-tinytex=0.57=r43hc72bb7e_0
  - r-truncnorm=1.0_9=r43h2b5f3a1_4
  - r-tzdb=0.5.0=r43h3697838_1
  - r-utf8=1.2.6=r43h2b5f3a1_0
  - r-uuid=1.2_1=r43hdb488b9_0
  - r-vctrs=0.6.5=r43h0d4f4ea_1
  - r-vioplot=0.5.1=r43hc72bb7e_0
  - r-viridislite=0.4.2=r43hc72bb7e_2
  - r-vroom=1.6.5=r43h0d4f4ea_1
  - r-withr=3.0.2=r43hc72bb7e_0
  - r-xfun=0.53=r43h3697838_0
  - r-xml=3.99_0.17=r43h5bae778_2
  - r-xml2=1.4.0=r43hc6fd541_0
  - r-xtable=1.8_4=r43hc72bb7e_6
  - r-yaml=2.3.10=r43hdb488b9_0
  - r-zoo=1.8_14=r43h2b5f3a1_0
  - raxml=8.2.13=h7b50bb2_3
  - readline=8.3=h853b02a_0
  - referencing=0.35.1=pyhd8ed1ab_0
  - requests=2.32.3=pyhd8ed1ab_0
  - rich=13.9.4=pyhd8ed1ab_0
  - rpds-py=0.20.0=py38h4005ec7_0
  - ruby=3.4.7=hd8161a3_0
  - salmon=1.10.3=haf24da9_3
  - samtools=1.22.1=h96c455f_0
  - scikit-learn=1.3.2=py38ha25d942_2
  - scipy=1.10.1=py38h59b608b_3
  - seaborn-base=0.13.2=pyhd8ed1ab_2
  - sed=4.9=h6688a6e_0
  - setuptools=75.3.0=pyhd8ed1ab_0
  - sip=6.7.12=py38h17151c0_0
  - six=1.16.0=pyh6c4a22f_0
  - slclust=02022010=h077b44d_6
  - snap=2017_03_01=h7b50bb2_0
  - sqlite=3.51.2=hbc0de68_0
  - statsmodels=0.14.1=py38h7f0c24c_0
  - stringtie=3.0.3=h29c0135_0
  - suitesparse=7.10.1=ha0f6916_7100102
  - sysroot_linux-64=2.28=h4ee821c_9
  - tantan=51=h5ca1c30_1
  - tar=1.35=h3b78370_0
  - tbb=2022.3.0=h8d10470_1
  - tbl2asn=25.7=h9ee0642_1
  - threadpoolctl=3.5.0=pyhc1e730c_0
  - tk=8.6.14=h39e8969_0
  - tktable=2.10=h8d826fa_7
  - toml=0.10.2=pyhd8ed1ab_0
  - tomli=2.0.2=pyhd8ed1ab_0
  - tomlkit=0.13.2=pyha770c72_0
  - transdecoder=5.7.1=pl5321hdfd78af_2
  - trimal=1.5.1=h9948957_0
  - trimmomatic=0.40=hdfd78af_0
  - trinity=2.15.2=pl5321h077b44d_6
  - trnascan-se=2.0.12=pl5321h7b50bb2_2
  - typing_extensions=4.12.2=pyha770c72_0
  - tzdata=2025c=hc9c84f9_1
  - ucsc-blat=482=hdc0a859_0
  - ucsc-fatotwobit=482=hdc0a859_0
  - ucsc-pslcdnafilter=482=h0b57e2e_0
  - ucsc-twobitinfo=482=hdc0a859_0
  - unicodedata2=15.1.0=py38h01eb140_0
  - urllib3=2.2.3=pyhd8ed1ab_0
  - wget=1.24.5=h251f7ec_0
  - wheel=0.45.1=pyhd8ed1ab_0
  - xcb-util=0.4.0=hd590300_1
  - xcb-util-image=0.4.0=h8ee46fc_1
  - xcb-util-keysyms=0.4.0=h8ee46fc_1
  - xcb-util-renderutil=0.3.9=hd590300_1
  - xcb-util-wm=0.4.1=h8ee46fc_1
  - xkeyboard-config=2.42=h4ab18f5_0
  - xlsxwriter=3.2.0=pyhd8ed1ab_0
  - xmltodict=0.14.2=pyhd8ed1ab_0
  - xorg-fixesproto=5.0=hb9d3cd8_1003
  - xorg-inputproto=2.3.2=hb9d3cd8_1003
  - xorg-kbproto=1.0.7=hb9d3cd8_1003
  - xorg-libice=1.1.2=hb9d3cd8_0
  - xorg-libsm=1.2.6=he73a12e_0
  - xorg-libx11=1.8.9=h8ee46fc_0
  - xorg-libxau=1.0.12=hb03c661_1
  - xorg-libxdmcp=1.1.5=hb03c661_1
  - xorg-libxext=1.3.4=h0b41bf4_2
  - xorg-libxfixes=5.0.3=h7f98852_1004
  - xorg-libxi=1.7.10=h4bc722e_1
  - xorg-libxrender=0.9.11=hd590300_0
  - xorg-libxt=1.3.0=hd590300_1
  - xorg-libxtst=1.2.5=h4bc722e_0
  - xorg-recordproto=1.14.2=hb9d3cd8_1003
  - xorg-renderproto=0.11.1=hb9d3cd8_1003
  - xorg-xextproto=7.3.0=hb9d3cd8_1004
  - xorg-xf86vidmodeproto=2.3.1=hb9d3cd8_1005
  - xorg-xproto=7.0.31=hb9d3cd8_1008
  - xz=5.8.1=hbcc6ac9_2
  - xz-gpl-tools=5.8.1=hbcc6ac9_2
  - xz-tools=5.8.1=hb9d3cd8_2
  - yaggo=1.5.10=0
  - yaml=0.2.5=h280c20c_3
  - yq=3.4.3=pyhd8ed1ab_0
  - zipp=3.21.0=pyhd8ed1ab_0
  - zlib=1.3.1=hb9d3cd8_2
  - zstandard=0.19.0=py38h0a891b7_0
  - zstd=1.5.7=hb78ec9c_6
```

</details>



You can then install this exact Conda environment by following:
```
cd ~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete
conda env create -n funannotate -f funannotate.yml
```

Installation (Non-FL users ):
**Remember this is not a requirement for Fisher Lab members, these steps have already been carried out for you.**

First, download databases for Funannotate:

**Update:** Original version did not include dbCAN, this has since been changed.

```
conda activate funannotate
funannotate setup -i merops -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i uniprot -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i pfam -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i go -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i mibig -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i interpro -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i gene2product -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i busco_outgroups -f -d ~/../projects/fisher-local-db/live/funannotate_db -w
funannotate setup -i repeats -f -d ~/../projects/fisher-local-db/live/funannotate_db -w
```
You will then need to update your database directory for funannotate
```
echo "export FUNANNOTATE_DB=~/../projects/fisher-local-db/live/funannotate_db" > ~/anaconda3/envs/tmp/etc/conda/activate.d/funannotate.sh
echo "unset FUNANNOTATE_DB" > ~/anaconda3/envs/tmp/etc/conda/deactivate.d/funannotate.sh
```
dbCAN failed to install database via funannotate. Therefore, the steps below outline how the database (version 12) was obtained manually:
```
conda activate funannotate
cd ~/../projects/fisher-local-db/live/funannotate_db
wget http://dbcan-hcc.unl.edu/download/Databases/V12/dbCAN-HMMdb-V12.txt 
wget http://dbcan-hcc.unl.edu/download/Databases/V12/CAZyDB.08062022.fam-activities.txt
wget http://dbcan-hcc.unl.edu/download/Databases/V12/readme.txt

mv CAZyDB.08062022.fam-activities.txt dbCAN-fam-HMMs.txt
mv readme.txt dbCAN.changelog.txt

sed 's/.hmm$//g' dbCAN-HMMdb-V12.txt > dbCAN.hmm
hmmpress -f dbCAN.hmm

echo -e 'dbCAN\thmmer3\t/rds/general/user/hchown/home/../projects/fisher-local-db/live/funannotate_db/dbCAN.hmm\t11.0\t2022-09-08\t750413\t4e8cafc3eea47ec7ba505bb1e3465d21' >> funannotate-db-info.txt
```
For GeneMark-ES/ET/EP, you'll need to download the tools to a local machine and move it to your desired folder see <http://topaz.gatech.edu/GeneMark/license_download.cgi> .
Once downloaded, you'll have to update the Perl script shebangs:
```
cd ~/../projects/fisher-local-software/live/genemark/gmes_linux_64_4
for file in *.pl
do
	sed -i 's|#!/usr/bin/perl|#!/usr/bin/env perl|g' $file
done
```
The main errors I obtained after this was from non-executable files in gmes. Double check the directory and make all the scripts executable. 
You will also need to make a copy of the gm_key and place that inside the directory.

We'll also download the databases for eggNOG:
```
cd ~/../projects/fisher-local-db/live/eggnog_db
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog.db.gz 
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog_proteins.dmnd.gz 
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog.taxa.tar.gz 

gunzip eggnog.db.gz 
gunzip eggnog_proteins.dmnd.gz 
tar -xvzf eggnog.taxa.tar.gz 
rm eggnog.taxa.tar.gz 
```
For antiSMASH databases, we will use the following:
```
conda activate antismash
mkdir ~/../projects/fisher-local-db/live/antismash
download-antismash-databases --database-dir ~/../projects/fisher-local-db/live/antismash
```
For Rfam databases (Infernal), we will use the following:
```
conda activate infernal
mkdir ~/../projects/fisher-local-db/live/rfam-15.1
cd ~/../projects/fisher-local-db/live/rfam-15.1
wget ftp://ftp.ebi.ac.uk/pub/databases/Rfam/15.1/Rfam.cm.gz
cmpress Rfam.cm 
```
There is also a conversion script that takes the table output of Infernal and converts this to GFF. I downloaded ```infernal-tblout2gff.pl``` from <https://github.com/nawrockie/jiffy-infernal-hmmer-scripts> and then placed it inside of the Fisher Lab local software directory.
```
cd ~/../projects/fisher-local-software/live
git clone https://github.com/nawrockie/jiffy-infernal-hmmer-scripts
```

Usage:

For your organism of interest, you will need to update ```BUSCO_DB``` and the seed-/species names in the ```predict``` command. 
You may even want to train your own gene prediction model. All of this information can be found in the Funannotate wiki (above).

Script for running basic gene prediction and annotation can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/annotate.job
```


```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=08:00:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zPredict
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish/ont.polished.fasta
ANNOT_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/annotation
FUNANNOTATE_DB=~/../projects/fisher-local-db/live/funannotate_db
BUSCO_DB=~/../projects/fisher-local-software/live/busco_downloads/lineages/eurotiales_odb10
GENEMARK_PATH=~/../projects/fisher-local-software/live/genemark/gmes_linux_64_4
STRAIN=demo
#--------------------------------------------
# 3. Sort and clean contig names
#--------------------------------------------
cd $TMPDIR

mkdir $ANNOT_DIR
BASE=$(basename $GENOME)
conda activate asm-qc
seqkit sort -l -r $GENOME | sed 's/>.*/>scaffold/g' | seqkit rename | sed 's/>scaffold$/>scaffold_1/g' | sed 's/>/>'"$STRAIN"'_/g' > $ANNOT_DIR/sorted.$BASE
#--------------------------------------------
# 4. Mask repetitive regions
#--------------------------------------------
conda activate funannotate
funannotate mask -i $ANNOT_DIR/sorted.$BASE --cpus 10 -o $ANNOT_DIR/masked.$BASE
#--------------------------------------------
# 5. Predict genes
#--------------------------------------------
# As we don't have corresponding RNA-seq data we will use a pre-trained database
# Estimated run time ~2hr45mins
mkdir $ANNOT_DIR/funannotate
export FUNANNOTATE_DB=$FUNANNOTATE_DB
export GENEMARK_PATH=$GENEMARK_PATH
funannotate predict -i $ANNOT_DIR/masked.$BASE -d $FUNANNOTATE_DB -o $ANNOT_DIR/funannotate --species "Aspergillus fumigatus" \
--strain $STRAIN --busco_seed_species aspergillus_fumigatus --cpus 10 --busco_db $BUSCO_DB --tmpdir $TMPDIR

#--------------------------------------------
# 6. Functional annotation
# Estimated run time ~4hrs
#--------------------------------------------
# InterProScan5
# Estimated run time 20mins
mkdir $ANNOT_DIR/interproscan
~/../projects/fisher-local-software/live/interproscan-5.67-99.0/interproscan.sh -appl Pfam --iprlookup -cpu 10 \
-d $ANNOT_DIR/interproscan --goterms -i "${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.proteins.fa"

# Eggnog
# Estimated run time ~2hrs
mkdir $ANNOT_DIR/eggnog
export EGGNOG_DATA_DIR=~/../projects/fisher-local-db/live/eggnog_db
emapper.py --sensmode very-sensitive --tax_scope Fungi --cpu 10 --output_dir $ANNOT_DIR/eggnog -o $STRAIN \
--override --temp_dir $TMPDIR -i "${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.proteins.fa"

# antiSMASH
# Estimated run time 30mins
conda activate antismash
mkdir $ANNOT_DIR/antismash
antismash --taxon fungi --clusterhmmer --tigrfam -c 10 \
--databases ~/../projects/fisher-local-db/live/antismash --output-dir $ANNOT_DIR/antismash --no-abort-on-invalid-records -d -v \
--genefinding-tool none \
"${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.gbk"

# Infernal
# Estimated run time ~1hr
conda activate infernal
mkdir $ANNOT_DIR/infernal
cmscan --rfam -E 0.001 --tblout $ANNOT_DIR/infernal/$STRAIN.cmscan --fmt 1 --cpu 10 \
~/../projects/fisher-local-db/live/rfam-15.1/Rfam.cm "${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.gbk"
~/../projects/fisher-local-software/live/jiffy-infernal-hmmer-scripts/infernal-tblout2gff.pl --cmscan \
$ANNOT_DIR/infernal/$STRAIN.cmscan > $ANNOT_DIR/infernal/$STRAIN.cmscan.gff

# Combine annotations (except Infernal)
conda activate funannotate
export FUNANNOTATE_DB=$FUNANNOTATE_DB
export BUSCO_DB=~/../projects/fisher-local-software/live/busco_downloads/lineages/eurotiales_odb10
funannotate annotate -i $ANNOT_DIR/funannotate/predict_results -d $FUNANNOTATE_DB \
-o $ANNOT_DIR/funannotate --fasta $ANNOT_DIR/masked.$BASE --species "Aspergillus fumigatus" \
--eggnog $ANNOT_DIR/eggnog/$STRAIN.emapper.annotations --antismash "${ANNOT_DIR}/antismash/Aspergillus_fumigatus_${STRAIN}.gbk" \
--iprscan "${ANNOT_DIR}/interproscan/Aspergillus_fumigatus_${STRAIN}.proteins.fa.xml" --busco_db $BUSCO_DB \
--cpus 10 --tmpdir $TMPDIR

```

**Line 26** organises the contigs/scaffolds in descending order and renames them as scaffold_1, scaffold_2...scaffold_N

**Line 31** performs soft-masking of repetitive regions. This helps reduces errors and speed-up performances of annotation tools. As a large fraction of the genome is repetitive in eukaryotes, if we filter out low-complexity regions we can focus on coding genes.

**Lines 38-41** run basic prediction of genes, mainly through the use of protein databases. See the Funannotate documentation to see the individual tools which comprise this step.

**Lines 50-51** annotate using InterProScan database of protein domains

**Lines 57-58** annotate using related orthologs

**Lines 64-67** annotates secondary metabolite gene clusters

**Lines 73-75** annotates RNA sequences

**Line 76** converts Infernal output to GFF

**Lines 80-86** perform additional annotations and combines results from InterProScan, EggNOG and antiSMASH

# Annotating transposable elements

## Basic TE-annotation

Tools for basic TE annotation:

* EarlGrey - TE annotater
  + Version: 18.8cc5c
  + [Paper](https://doi.org/10.1038/nbt.3988) 
  + [GitHub](https://github.com/soedinglab/mmseqs2)

Installation:
```
conda create -n earlgrey bioconda::earlgrey
conda activate earlgrey
earlGrey --help
```
You will then notice a Bash script will be generated for you to install the relevant Dfam partitiions <configure_dfam39.sh>.
Downloading the full set will cost you a lot of memory so I recommend (for fungi) to only use 0 and 6.
Therefore, open the file and change this line
```
curl -o 'dfam39_full.#1.h5.gz' 'https://dfam.org/releases/current/families/FamDB/dfam39_full.[0-16].h5.gz'
```
to:
```
curl -o 'dfam39_full.16.h5.gz' 'https://dfam.org/releases/current/families/FamDB/dfam39_full.16.h5.gz'
curl -o 'dfam39_full.0.h5.gz' 'https://dfam.org/releases/current/families/FamDB/dfam39_full.0.h5.gz'
```
Then run, as per:
```
bash configure_dfam39.sh
```
Usage:


Script for annotating TEs can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/earlgrey.job
```

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=01:30:00
#PBS -l select=1:ncpus=30:mem=80gb
#PBS -N zEarlgrey
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/annotation/sorted.ont.polished.fasta
OUTDIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/annotation/earlgrey
PREFIX=demo
#--------------------------------------------
# 3. Annotate TEs
#--------------------------------------------
conda activate earlgrey
mkdir -p $OUTDIR
cd $TMPDIR
mkdir earlgrey_raw
OMP_NUM_THREADS=30
earlGrey -g $GENOME -t 16 -s $PREFIX -o earlgrey_raw  -r fungi -c yes -d yes -e yes
cp -r ./earlgrey_raw/${PREFIX}_EarlGrey $OUTDIR
```




## Identify RIP-sites

RIPper
https://doi.org/10.7717/peerj.7447
https://github.com/TheRIPper-Fungi/TheRIPper


To identify sites of RIP we can use a web-app called "The RIPper".
Here we can upload FASTA files containing whole-genomes or contigs and perform a RIP analysis. 
The resulting GFF3 file can then be downloaded and used to compare against TE sites. 
The web-page has a brief introduction on how to use the site.
For our purposes, we will use default parameters:
Files -> Upload File -> RIP Genome -> Export GFF3 RIP data (Genome)

# Quick pangenome generation

A pangenome in its crudest form is the total pool of genes across a species.
This definition can be expanded in many ways to including coding/non-coding genes or to incorporate multiple species that full under a taxonomic group.
In this example, we are focusing on within-species and coding genes only.
The approach used simply clusters protein sequences based on identity (90%) and coverage (90%), using a single tool (MMseqs2).
Downstream processing then uses Bash scripting (with thanks to ChatGPT) to convert the results into a more traditional "pangenome" structure.
This is essentially a matrix with orthogroups (collections of similar genes) as the rows and strain ID's as the columns.
Alongside this, cluster-gene and gene-cluster keys/edge-lists have been generated along with binary and numeric matrix representations.
By generating a pangenome, we are able to understand the differences at the gene-level and provide a source of homology to root synteny analyses.

Tools for clustering:

* MMseqs2 - sequence clustering
  + Version: 18.8cc5c
  + [Paper](https://doi.org/10.1038/nbt.3988) 
  + [GitHub](https://github.com/soedinglab/mmseqs2)


Installation:
```
conda create -n mmseqs2 bioconda::mmseqs2

```
Usage:

To generate a pangenome, we use a file containing the realpaths of amino acid FASTA files.
We highly recommend adding 1-2 reference proteomes so that you can compare against a ground truth (or close enough).
For FL members, we have a demonstration list of protein paths, however, the user name must be changed respectively.

Script for generating a quick pangenome can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/pangenome.job
```

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:20:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zPangenome
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
OUTDIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/pangenome
LIST_OF_PEP=/rds/general/user/hchown/projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/protein_list.txt
#--------------------------------------------
# 3. Generate a database of peptide sequences
#--------------------------------------------
mkdir -p $OUTDIR
cd $TMPDIR
cp -t $TMPDIR $(cat $LIST_OF_PEP)
# Add unique ID to FASTA headers
for file in *
do
	PREFIX=$(basename $file | cut -d"." -f1)
	sed -i 's/>/>'"$PREFIX"'@/g' $file
done
# Combine files
cat * > total.faa
#--------------------------------------------
# 4. Cluster peptides
#--------------------------------------------
conda activate mmseqs2
# Cluster
mmseqs easy-cluster total.faa $OUTDIR/clusterRes tmp --min-seq-id 0.9 -c 0.9 --cov-mode 1 --threads 10
#--------------------------------------------
# 5. Generate simple match tables
# Cluster - Gene and Gene - Cluster
#--------------------------------------------
awk '
{
  if (!seen[$1]) {
    seen[$1] = ++c
  }
  printf "OG%06d\t%s\n", seen[$1], $0
}
' $OUTDIR/clusterRes_cluster.tsv | cut -f1,3  > $OUTDIR/cluster_gene.tsv

awk '{print $2,$1}' $OUTDIR/cluster_gene.tsv > $OUTDIR/gene_cluster.tsv
#--------------------------------------------
# 6. Create a matrix with gene IDs
#--------------------------------------------
gawk -F'\t' '
BEGIN { OFS="\t" }

{
  og = $1
  split($2, a, "@")
  strain = a[1]
  gene   = a[2]

  strains[strain] = 1
  ogs[og] = 1

  key = og SUBSEP strain
  if (cell[key] == "") cell[key] = gene
  else                 cell[key] = cell[key] "," gene
}

END {
  # sort strains and OGs
  nS = asorti(strains, S)
  nO = asorti(ogs, O)

  # header
  printf "Cluster"
  for (i=1; i<=nS; i++) printf OFS S[i]
  print ""

  # body
  for (i=1; i<=nO; i++) {
    og = O[i]
    printf "%s", og
    for (j=1; j<=nS; j++) {
      strain = S[j]
      key = og SUBSEP strain
      printf OFS (key in cell ? cell[key] : "")
    }
    print ""
  }
}
' $OUTDIR/cluster_gene.tsv > $OUTDIR/cluster_matrix.gene_name.tsv
#--------------------------------------------
# 7. Convert to count format
#--------------------------------------------
gawk -F'\t' '
BEGIN{OFS="\t"}
NR==1 {print; next}
{
  printf "%s", $1
  for (i=2; i<=NF; i++) {
    if ($i=="" ) c=0
    else { c = split($i, a, ",") }   # number of comma-separated genes
    printf OFS c
  }
  print ""
}
' $OUTDIR/cluster_matrix.gene_name.tsv > $OUTDIR/cluster_matrix.count.tsv
#--------------------------------------------
# 8. Conver to binary format
#--------------------------------------------
# Convert to binary
gawk -F'\t' '
BEGIN{OFS="\t"}
NR==1 {print; next}
{
  printf "%s", $1
  for (i=2; i<=NF; i++) {
    b = ($i+0 > 0) ? 1 : 0
    printf OFS b
  }
  print ""
}
' $OUTDIR/cluster_matrix.count.tsv > $OUTDIR/cluster_matrix.binary.tsv


```
**Line 13** supplies a list of peptide file paths

**Line 19** makes a copy of these peptide files in your temporary space

**Lines 21-27** adds a strain ID (from the name of the input file) to each FASTA header and combines peptide files.
This enables us to figure out which strain each gene is coming from as annotation tools use re-occuring gene IDs usually based on order of processing.

**Line 33** clusters peptide sequences using a minimum sequence identity of 90% and coverage of 90%, creating a highly similar pangenome

**Lines 38-47** reformats the cluster output to create a orthogroup ID ranging from 1...N clusters. 
By doing so, we are generating a tidier, universal, cluster name. 
We then repeat this to generate both a cluster-gene and gene-cluster key

**Lines 51-90** creates a matrix whereby rows indicate clusters and columns are individuals (using the "@" symbol as a delimiter).
If multiple genes from the same individual have clustered together, these are displayed as comma-separated lists

**Lines 94-106** creates a numeric count matrix, counting the number of clustered genes per orthogroup/cluster

**Lines 110-122** converts this numeric matrix into a binary matrix, with 1 denoting presence and 0 absence

# 5mC and 5hmC from ONT

In this example, we are aiming to identify 5mC methylation sites from ONT data. 5mC sites are just one form of epigenetic markers; ONT can sequence alternative markers depending on the software which performs the base-calling, additional markers can be identified if you adapt the code below.


Installation (non-/FL):
```
conda create -n ont-meth bioconda::ont-modkit 
```

Installation (non-FL):

Fisher Lab members do not need to carry out these steps
```
cd ~/../projects/fisher-local-software/live
curl "https://cdn.oxfordnanoportal.com/software/analysis/dorado-1.3.0-linux-x64.tar.gz" -o dorado-1.3.0-linux-x64.tar.gz
tar -xzf dorado-1.3.0-linux-x64.tar.gz
```

In this example, we will be using demonstration data generated by an MRes student (Peng Zhou) during their time in Fisher Lab. 
For non-FL members please adapt the scripts appropriately.

Two scripts have been generated to call 5mC sites. The first, ```01_methAlign.job```, takes unaligned ```.bam``` files and aligns these to their respective genome. This is carried out independently for each file generated for your ONT sample. The second, ```02_methCall.job```, combines these files together, identifies 5mC and 5hmC sites, filters these based on a minimum percentage of modified base calls (75%) and identifies all of the sites, plus those on CpG dinucleotides. Much like variant calling, reads are overlapped onto eachother, this means that a portion of the reads may contain a modified base, whilst others do not, this is due to repetitive nature of DNA and the overlaps generated during assembly. This is why we apply an additional filter on percentage of modified bases.

## Alignment

Tools for aligning ONT BAM files:

* Dorado - ONT toolkit
  + Version: 1.3.0
  + [GitHub](https://github.com/nanoporetech/dorado) 
* SAMtools, see details above

Usage:

Script for aligning BAM files from ONT can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/01_methAlign.job
```

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=04:00:00
#PBS -l select=1:ncpus=10:mem=80gb
#PBS -N zMethAlign
#PBS -J 1-50
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
BAM_FILE=$(sed -n "${PBS_ARRAY_INDEX}p" /rds/general/user/hchown/projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/bam_file_list.txt)
RUN_ID=$(echo "$BAM_FILE" | rev | cut -d"/" -f2 | rev)
FASTA_DIR=~/../ephemeral/
BAM_OUTDIR=~/../ephemeral/$RUN_ID/bam
#--------------------------------------------
# 3. Align BAM to genome
#--------------------------------------------
cd $TMPDIR
conda activate polish
mkdir -p $BAM_OUTDIR
~/../projects/fisher-local-software/live/dorado-1.3.0-linux-x64/bin/dorado aligner -t 10 $FASTA_DIR/$RUN_ID.ont.polished.fasta $BAM_FILE \
| samtools sort --threads 10 > $BAM_OUTDIR/$RUN_ID.$PBS_ARRAY_INDEX.aligned_reads.bam
```

**Lines 23-24** align the ONT BAM file to the genome and sorts the resulting file

## Calling modifications

Tools for aligning ONT BAM files:

* Modkit - ONT epi-marker caller
  + Version: 1.8.17
  + [GitHub](https://github.com/nanoporetech/modkit) 
* SAMtools, see details above

Usage:

Script for calling 5hmC and 5mC from aligned BAM files can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/01_methAlign.job
```

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:20:00
#PBS -l select=1:ncpus=10:mem=80gb
#PBS -N zMethCall
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
BAM_DIR=~/../ephemeral/barcode01/bam
GENOME=~/../ephemeral/barcode01.ont.polished.fasta
OUT_DIR=~/../ephemeral
#--------------------------------------------
# 3. Combine BAM alignments
#--------------------------------------------
cd $TMPDIR
conda activate polish
find $BAM_DIR -name "*.bam" -print0 | xargs -0 samtools merge -@ 10 barcode01.merged.bam
samtools index barcode01.merged.bam
#--------------------------------------------
# 4. Sanity check methylation sites
#--------------------------------------------
conda activate ont-meth
modkit modbam summary --filter-threshold 0.79 --no-sampling -t 10 barcode01.merged.bam
#--------------------------------------------
# 5. Call methylation sites
#--------------------------------------------
modkit pileup \
barcode01.merged.bam \
barcode01.meth.bed \
--modified-bases 5mC 5hmC \
--reference $GENOME \
--log ~/../ephemeral/modkit.log.txt \
-t 10 \
--filter-threshold 0.79
# CpG sites only
modkit pileup \
barcode01.merged.bam \
barcode01.meth.cpg.bed \
--modified-bases 5mC 5hmC \
--reference $GENOME \
--log ~/../ephemeral/modkit.log.txt \
-t 10 --cpg \
--filter-threshold 0.79
#------------------------------------------
# 6. Filter sites
#-----------------------------------------
awk -F '\t' '$11 > 75' barcode01.meth.bed > $OUT_DIR/barcode01.filtered.meth.bed
awk -F '\t' '$11 > 75' barcode01.meth.cpg.bed > $OUT_DIR/barcode01.filtered.meth.cpg.bed

```

**Lines 20-21** merge the individual BAM files from a barcode and indexes the resulting file

**Line 26** prints out a brief summary of the modification profiles

**Lines 30-37** calls 5mC and 5hmC modified bases, filter threshold has been set to 0.79

**Lines 39-46** identifies the 5mC and 5hmC sits that are on CpG islands

**Lines 50-51** filters the calls so that the modified base must appear in more than 75% of the base calls


