---
title: "Genome Assembly and Annotation"
author: "Harry Chown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    highlight: pygments
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction to genome assembly and annotation



## Caveats

The aim of this tutorial is teach you the process of assembling and annotating your first fungal genome. The methodology outlined below has been specifically designed to annotate the genome of *A. fumigatus* isolates. Modifications to this tutorial will enable genome analyses of alternative fungal organisms, but please do ensure to read the original software documentation to identify where those changes will be required. I will attempt to notify any specific parameters that I use for *A. fumigatus* but I may have missed some.

Additionally, this tutorial may change over time to include assembling from various sequencing platforms and incorporation of new analysis tools. However, older versions will be available through releases.

Furthermore, the code and scripts outlined in this tutorial has been design for use on ICL HPC systems. They follow a simple format of job scripts which can be deployed on this system. Installation guidance is also provided to help members of ICL install the relevant tools on this system. This may mean that this is not 100% transferable to the system you are working on. In addition to this, folder structure will be relevant to members working within Matthew Fisher's group, so check and update the relevant file/folder paths when required. Please seek guidance elsewhere if installation and running of tools fails when used on your relevant system.

## Demonstration data

### Short-read data
Paired-end short-read sequence data derived from *A. fumigatus* strain CEA10 using NextSeq 500 under the accession [SRR12849133](https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR12849133). Members of FisherLab can find data at:
```
~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_*.fastq.gz
```
### ONT long-read data
Long-read sequence data derived from *A. fumigatus* strain CEA10 using ONT under the accession [SRR28036069](https://trace.ncbi.nlm.nih.gov/Traces/?run=SRR28036069). Members of FisherLab can find data at:
```
~/../projects/fisher-aspergillus-rawdata/live/tutorial/SRR28036069.fastq.gz
```
# Read quality control (QC)

## Long-read QC
We are going to generate a conda environment called ```ont-qc``` which contains three tools in order to assess the quality of ONT reads, trim and filter. The three tools are:

* NanoQC - visualise quality
  + Version: 0.10.0
  + [GitHub](https://github.com/wdecoster/nanoQC) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149)
* NanoFilt - filter reads
  + Version: 2.8.0
  + [GitHub](https://github.com/wdecoster/nanofilt)
  + [Paper](https://doi.org/10.1093/bioinformatics/bty149) (same as NanoQC)
* porechop - trim reads
  + Version: 0.2.4
  + [GitHub](https://github.com/rrwick/Porechop)
  + [Paper](https://doi.org/10.1099/mgen.0.000132)

Installation:
```
conda create -n ont-qc bioconda::nanoqc bioconda::porechop bioconda::nanofilt
```
Script for ONT read QC analysis can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_qc.job
```
Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:40:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zONT_qc
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
INPUT=~/../projects/fisher-aspergillus-rawdata/live/tutorial/SRR28036069.fastq.gz
QC_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data
#--------------------------------------------
# 3. QC long-reads
#CPU Time used: 02:54:36
#CPU Percent: 975%
#Memory usage: 6260156kb
#Approx Power usage: 0.0348
#Walltime usage: 00:29:40
#--------------------------------------------
conda activate ont-qc
# Initial QC
cd $TMPDIR
mkdir -p $QC_DIR/ONT/initial_QC
nanoQC -o $QC_DIR/ONT/initial_QC $INPUT
gunzip -c $INPUT > SRR28036069.fastq
# Read trimming and filtering
porechop -i SRR28036069.fastq -t 10 --format fastq | \
NanoFilt -q 10 -l 500  > $QC_DIR/ONT/SRR28036069.trimmed.fastq
# Post-filter QC
mkdir -p $QC_DIR/ONT/post_QC
nanoQC -o $QC_DIR/ONT/post_QC $QC_DIR/ONT/SRR28036069.trimmed.fastq
```
**Lines 12 and 13** sets the desired input files and output directory.

**Lines 26 and 33** visualse QC on reads before and after filtering, respectively

**Lines 29 and 30** trims using porechop's default settings and filters (NanoFilt) based on average read quality >10 or length <500bp. 



# Assembly

## Long-read assembly

### ONT assembly

#### Initial assembly
We are going to use a single tool to generate our initial genome assembly:

* Flye - long-read assembler
  + Version: 2.9.6-b1802
  + [GitHub](https://github.com/mikolmogorov/Flye/tree/flye) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n flye bioconda::flye
```
Script for ONT read assembly can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm.job
```

Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:40:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zONT_asm
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
INPUT=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/SRR28036069.trimmed.fastq
ASM_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/flye
#--------------------------------------------
# 3. Assemble ONT reads
# CPU Time used: 01:32:43
# CPU Percent: 968%
# Memory usage: 10680116kb
# Approx Power usage: 0.0184
# Walltime usage: 00:14:19
#--------------------------------------------
mkdir $ASM_DIR
conda activate flye
# Initial assembly
flye --nano-hq $INPUT -o $ASM_DIR --threads 10
```
**Line 25** assemble long-reads

### Improvement with short-read data

Tools for alignment, indexing and polishing:

* Minimap2 - sequence alignment
  + Version: 2.30-r1287
  + [GitHub](https://github.com/lh3/minimap2) 
  + [Paper](https://doi.org/10.1093/bioinformatics/bty191)
* SAMtools - alignment manipulation
  + Version: 1.23
  + [GitHub](https://github.com/samtools/samtools) 
  + [Paper](https://doi.org/10.1093/gigascience/giab008)
* Pilon - genome polisher
  + Version: 1.24
  + [GitHub](https://github.com/broadinstitute/pilon) 
  + [Paper](https://doi.org/10.1038/s41587-019-0072-8)
  
Installation:
```
conda create -n polish bioconda::minimap2 bioconda::samtools
```
Additional installation was required for Pilon. **Fisher Lab users do not need to repeat the following steps.**
```
cd ~/../projects/fisher-local-software/live
mkdir pilon
cd pilon
wget https://github.com/broadinstitute/pilon/releases/download/v1.24/pilon-1.24.jar
```

Script for polishing assemblies with short-read data can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/polish.job
```

Usage:
```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=01:00:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zPolish
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/flye/assembly.fasta
R1=~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_1.fastq.gz
R2=~/../projects/fisher-aspergillus-rawdata/live/tutorial/UU_2.fastq.gz
POLISH_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish
PILON_PATH=~/../projects/fisher-local-software/live/pilon/pilon-1.24.jar
#--------------------------------------------
# 3. Generate polish rounds
#CPU Time used: 00:41:08
#CPU Percent: 168%
#Memory usage: 12210052kb
#Approx Power usage: 0.008199999999999999
#Walltime usage: 00:35:53
#--------------------------------------------
mkdir $POLISH_DIR
conda activate polish
cd $TMPDIR
cp $GENOME ./tmp.fasta
for i in $(seq 1 5); do
	mkdir pilon_tmp
	# Align sequence data to genome
	minimap2 -ax sr tmp.fasta $R1 $R2 | samtools sort -o output.sorted.bam
	samtools index output.sorted.bam
	# Use pilon to improve genome based on alignment
	java -Xmx100G -jar $PILON_PATH --genome tmp.fasta --bam output.sorted.bam \
	--outdir pilon_tmp --changes 
	# Overwrite genome 
	mv pilon_tmp/pilon.fasta tmp.fasta
	rm -rf pilon_tmp
done

cp tmp.fasta $POLISH_DIR/ont.polished.fasta

```
**Lines 30 and 40** establishes a loop to repeat the polish 5 times

**Line 33 and 34** aligns reads to the genome, sorts and index alignment file

**Lines 36 and 37** improves the assembly using the mapped data

**Lines 38 and 39** overwrites the input file with the new output in order to run polishing again

**Line 42** saves the results in the respective output directory

# Assembly QC
Tools for quality control of assemblies:

* SeqKit - basic genome statistics
  + Version: 2.12.0
  + [GitHub](https://github.com/shenwei356/seqkit) 
  + [Paper](https://doi.org/10.1002/imt2.191)
* BUSCO - gene content of near-universal single-copy orthologs
  + Version: 6.0.0
  + [Website](https://busco.ezlab.org/) 
  + [Paper]( https://doi.org/10.1093/nar/gkae987)
* QualiMap - coverage analyser
  + Version: 2.3
  + [Website](http://qualimap.conesalab.org/) 
  + [Paper]( https://doi.org/10.1093/bioinformatics/btv566)
* Minimap2, SAMtools - for read alignments to assembly, see details above

Installation:
We will also be using minimap2 and samtools which can be found in the ```polish``` Conda environment. Installation is outlined above.
Additionally, we will be running an offline BUSCO search using the appropriate database (eurotiales) for *A. fumigatus*. Details on downloading the appropriate BUSCO database can be found at: <https://busco.ezlab.org/>.
```
conda create -n asm-qc bioconda::busco=6.0.0 bioconda::sepp=4.5.5 bioconda::qualimap bioconda::seqkit
```


Note: if assembling an organism that is not part of the *Aspergilli* you may need to update the BUSCO database; if running on an alternative machine you will need to change to the appropriate download path. Additionally, read mapping is only for long-read sequence data, please see section on polishing to edit Minimap2 for short-read data.

Script for running assembly QC (derived from long-read data) can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/ONT_asm_qc.job
```


Usage:

```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=00:20:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zAsmQC
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish/ont.polished.fasta
ONT=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/SRR28036069.trimmed.fastq
BUSCO_DB=~/../projects/fisher-local-software/live/busco_downloads
ASM_QC_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/asm_QC
#--------------------------------------------
# 3. Perform QC on polished assembly
#--------------------------------------------
mkdir $ASM_QC_DIR
conda activate asm-qc
# Traditional statistics
seqkit stats -a -N 50,90 $GENOME | sed 's/,//g' > $ASM_QC_DIR/basic_stats.txt
# Conserved universal orthologs
cd $TMPDIR
mkdir busco
busco -m genome -i $GENOME -o assembly --out_path busco -l eurotiales_odb10 \
--miniprot --offline \
--download_path ~/../projects/fisher-local-software/live/busco_downloads \
-f -c 10
cp ./busco/assembly/short_summary.specific.eurotiales_odb10.assembly.txt $ASM_QC_DIR/busco.txt
# Coverage 
## ONT read mapping
conda activate polish
minimap2 -a -x map-ont $GENOME $ONT | samtools sort -T tmp -o output.sorted.bam
samtools index output.sorted.bam
## Coverage calc.
conda activate asm-qc
mkdir $ASM_QC_DIR/qualimap
qualimap bamqc -nt 10 -bam output.sorted.bam -outdir $ASM_QC_DIR/qualimap \
--java-mem-size=80G
#--------------------------------------------
# 4. Summarise QC results
#--------------------------------------------
#num_seqs sum_len min_len avg_len max_len median GC% N50 N90
STATS=$(awk '{print $4,$5,$6,$7,$8,$10,$18,$20,$21'} $ASM_QC_DIR/basic_stats.txt \
| tail -n1 | sed 's/ /,/g')
#BUSCO Complete	| BUSCO Single |	BUSCO Duplicate	 | BUSCO Fragmented | 	BUSCO Missing	| BUSCO Num Genes
BUSCO=$(grep "C:" $ASM_QC_DIR/busco.txt | awk '{print $1}' | sed 's/\[/,/g' \
| sed 's/\]//g' | sed 's/[%|:|Aa-Zz]//g' | cut -d"," -f1-6)
COV=$(cat $ASM_QC_DIR/qualimap/genome_results.txt | grep "mean coverageData" \
| cut -d"=" -f2 | sed 's/ //g' | sed 's/X//g')
echo "${STATS},${COV},${BUSCO}" > $ASM_QC_DIR/overall_stats.csv
```
**Line 22** performs basic calculation of assembly statistics

**Lines 26-29** identifies the presence of single-copy conserved orthologs.
This provides an idea of genome completeness

**Lines 33-40** aligns reads to the genome (minimap2, samtools) and calculates
coverage (qualimap), providing an idea of read-depth

**Lines 45-52** extract formatted information from outputs and saves this as a 
single comma-separated line. Exact headers are described a combination of 
**Line 44 + coverage + Line 47**

# Basic gene prediction and annotation

Gene prediction and annotation are two major steps in understanding the constituent elements of a genome. Here we conjoin the two stages: prediction (identifying where genes are present) and annotation (characterizing identified genes). It's important to realize that these are two distinct steps. 

The pipeline below predominantly uses Funannotate and the tutorials from the respective wiki page. You will notice that many of the annotation steps we perform outside of Funannotate and combine at the end. As there are many steps in this script, please check the error/output files to troubleshoot problems as they arise.

Installation of components here was somewhat tricky, however I have tried to outline as much as I can for non-Fisher Lab (FL) users. Thus, I have separated out the installation segments for FL and non-FL users. It is important to note the database versions used, as they could impact your results.

Tools for prediction and annotation:

* Funannotate - fungal gene prediction
  + Version: 1.8.17
  + [GitHub](https://github.com/nextgenusfs/funannotate) 
  + [Wiki](http://funannotate.readthedocs.io/)
* InterProScan5 - annotation from various protein domain databases
  + Version: 5.67-99.0
  + [Wiki](https://interproscan-docs.readthedocs.io/en/v5/) 
  + [Paper](https://doi.org/10.1093/nar/gkaa977)
* EggNOG - functional annotation from orthologs
  + Version: 2.1.13
  + [GitHub](https://github.com/eggnogdb/eggnog-mapper) 
  + [Paper](https://doi.org/10.1093/molbev/msab293)
* AntiSMASH - secondary metabolite gene clusters
  + Version: 8.0.4
  + [GitHub](https://github.com/antismash/antismash) 
  + [Paper](https://doi.org/10.1093/nar/gkaf334)
* Infernal - RNA annotation
  + Version: 1.1.5
  + [Website](http://eddylab.org/infernal/) 
  + [Paper](https://doi.org/10.1093/bioinformatics/btt509)
* Seqkit, see details above

Installation (Non-/FL users):
```
conda create -n funannotate "python>=3.6,<3.9" bioconda::funannotate bioconda::eggnog-mapper
conda create -n antismash bioconda::antismash
conda create -n infernal bioconda::infernal
```

Installation (Non-FL users ):
Remember this is not a requirement for Fisher Lab members, these steps have already been carried out for you.

First, download databases for Funannotate:

At the time of writing this (Jan. 2025) access to dbCAN databases was not possible and has therefore been omitted.
```
conda activate funannotate
funannotate setup -i merops -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i uniprot -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i pfam -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i go -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i mibig -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i interpro -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i gene2product -f -d ~/../projects/fisher-local-db/live/funannotate_db
funannotate setup -i busco_outgroups -f -d ~/../projects/fisher-local-db/live/funannotate_db -w
funannotate setup -i repeats -f -d ~/../projects/fisher-local-db/live/funannotate_db -w
```
For GeneMark-ES/ET/EP, you'll need to download the tools to a local machine and move it to your desired folder see <http://topaz.gatech.edu/GeneMark/license_download.cgi> .
Once downloaded, you'll have to update the Perl script shebangs:
```
cd ~/../projects/fisher-local-software/live/genemark/gmes_linux_64_4
for file in *.pl
do
	sed -i 's|#!/usr/bin/perl|#!/usr/bin/env perl|g' $file
done
```
The main errors I obtained after this was from non-executable files in gmes. Double check the directory and make all the scripts executable. 
You will also need to make a copy of the gm_key and place that inside the directory.

We'll also download the databases for eggNOG:
```
cd ~/../projects/fisher-local-db/live/eggnog_db
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog.db.gz 
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog_proteins.dmnd.gz 
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.2/eggnog.taxa.tar.gz 

gunzip eggnog.db.gz 
gunzip eggnog_proteins.dmnd.gz 
tar -xvzf eggnog.taxa.tar.gz 
rm eggnog.taxa.tar.gz 
```
For antiSMASH databases, we will use the following:
```
conda activate antismash
mkdir ~/../projects/fisher-local-db/live/antismash
download-antismash-databases --database-dir ~/../projects/fisher-local-db/live/antismash
```
For Rfam databases (Infernal), we will use the following:
```
conda activate infernal
mkdir ~/../projects/fisher-local-db/live/rfam-15.1
cd ~/../projects/fisher-local-db/live/rfam-15.1
wget ftp://ftp.ebi.ac.uk/pub/databases/Rfam/15.1/Rfam.cm.gz
cmpress Rfam.cm 
```


Usage:

For your organism of interest, you will need to update ```BUSCO_DB``` and the seed-/species names in the ```predict``` command. 
You may even want to train your own gene prediction model. All of this information can be found in the Funannotate wiki (above).

Script for running basic gene prediction and annotation can be found at:
```
~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/complete/annotate.job
```


```bash {.numberLines}
#!/bin/bash
#PBS -l walltime=08:00:00
#PBS -l select=1:ncpus=10:mem=100gb
#PBS -N zPredict
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
GENOME=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/polish/ont.polished.fasta
ANNOT_DIR=~/../projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/data/ONT/annotation
FUNANNOTATE_DB=~/../projects/fisher-local-db/live/funannotate_db
BUSCO_DB=~/../projects/fisher-local-software/live/busco_downloads/lineages/eurotiales_odb10
GENEMARK_PATH=~/../projects/fisher-local-software/live/genemark/gmes_linux_64_4
STRAIN=demo
#--------------------------------------------
# 3. Sort and clean contig names
#--------------------------------------------
cd $TMPDIR
rm -rf $ANNOT_DIR
mkdir $ANNOT_DIR
BASE=$(basename $GENOME)
conda activate asm-qc
seqkit sort -l -r $GENOME | sed 's/>.*/>scaffold/g' | seqkit rename | sed 's/>scaffold$/>scaffold_1/g'  > $ANNOT_DIR/sorted.$BASE
#--------------------------------------------
# 4. Mask repetitive regions
#--------------------------------------------
conda activate funannotate
funannotate mask -i $ANNOT_DIR/sorted.$BASE --cpus 10 -o $ANNOT_DIR/masked.$BASE
#--------------------------------------------
# 5. Predict genes
#--------------------------------------------
# As we don't have corresponding RNA-seq data we will use a pre-trained database
# Estimated run time ~2hr45mins
mkdir $ANNOT_DIR/funannotate
export FUNANNOTATE_DB=$FUNANNOTATE_DB
export GENEMARK_PATH=$GENEMARK_PATH
funannotate predict -i $ANNOT_DIR/masked.$BASE -d $FUNANNOTATE_DB -o $ANNOT_DIR/funannotate \
--species "Aspergillus fumigatus" --strain $STRAIN \
--busco_seed_species aspergillus_fumigatus --cpus 10 \
--busco_db $BUSCO_DB --tmpdir $TMPDIR
#--------------------------------------------
# 6. Functional annotation
# Estimated run time ~4hrs
#--------------------------------------------
# InterProScan5
# Estimated run time 20mins
mkdir $ANNOT_DIR/interproscan
~/../projects/fisher-local-software/live/interproscan-5.67-99.0/interproscan.sh -appl Pfam \
--iprlookup -cpu 10 -d $ANNOT_DIR/interproscan --goterms \
-i "${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.proteins.fa"

# Eggnog
# Estimated run time ~2hrs
mkdir $ANNOT_DIR/eggnog
export EGGNOG_DATA_DIR=~/../projects/fisher-local-db/live/eggnog_db
emapper.py --sensmode very-sensitive --tax_scope Fungi --cpu 10 \
--output_dir $ANNOT_DIR/eggnog -o $STRAIN --override --temp_dir $TMPDIR \
-i "${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.proteins.fa"

# antiSMASH
# Estimated run time 30mins
conda activate antismash
mkdir $ANNOT_DIR/antismash
antismash --taxon fungi --clusterhmmer --tigrfam -c 10 \
--databases ~/../projects/fisher-local-db/live/antismash \
--output-dir $ANNOT_DIR/antismash --no-abort-on-invalid-records -d -v \
--genefinding-tool none \
"${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.gbk"

# Infernal
# Estimated run time ~1hr
conda activate infernal
mkdir $ANNOT_DIR/infernal
cmscan --rfam -E 0.001 --tblout $ANNOT_DIR/infernal/$STRAIN.cmscan --fmt 1 --cpu 10 \
~/../projects/fisher-local-db/live/rfam-15.1/Rfam.cm "${ANNOT_DIR}/funannotate/predict_results/Aspergillus_fumigatus_${STRAIN}.gbk"
# dbCAN database is currently down for maintenance and is not possible to run

# Combine annotations (except Infernal)
conda activate funannotate
funannotate annotate -i $ANNOT_DIR/funannotate/predict_results -d $FUNANNOTATE_DB \
-o $ANNOT_DIR/funannotate --fasta $ANNOT_DIR/masked.$BASE --species "Aspergillus fumigatus" \
--eggnog $ANNOT_DIR/eggnog/$STRAIN.emapper.annotations \
--antismash "${ANNOT_DIR}/antismash/Aspergillus_fumigatus_${STRAIN}.gbk" \
--iprscan "${ANNOT_DIR}/interproscan/Aspergillus_fumigatus_${STRAIN}.proteins.fa.xml" \
--busco_db $BUSCO_DB \
--cpus 10 --tmpdir $TMPDIR
```

**Line 26** organises the contigs/scaffolds in descending order and renames them as scaffold_1, scaffold_2...scaffold_N

**Line 31** performs soft-masking of repetitive regions. This helps reduces errors and speed-up performances of annotation tools. As a large fraction of the genome is repetitive in eukaryotes, if we filter out low-complexity regions we can focus on coding genes.

**Lines 38-43** run basic prediction of genes, mainly through the use of protein databases. See the Funannotate documentation to see the individual tools which comprise this step.

**Lines 51-53** annotate using InterProScan database of protein domains

**Lines 59-61** annotate using related orthologs

**Lines 67-71** annotates secondary metabolite gene clusters

**Lines 77-78** annotates RNA sequences

**Lines 83-89** perform additional annotations and combines results from InterProScan, EggNOG and antiSMASH

# Annotating methylation sites from ONT data
Installation:
```
cd ~/../projects/fisher-local-software/live
curl "https://cdn.oxfordnanoportal.com/software/analysis/dorado-1.3.0-linux-x64.tar.gz" -o dorado-1.3.0-linux-x64.tar.gz
tar -xzf dorado-1.3.0-linux-x64.tar.gz
~/../projects/fisher-local-software/live/dorado-1.3.0-linux-x64/bin/dorado --version

conda create -n ont-meth bioconda::ont-modkit 
```
Usage:
Mapping
```
#!/bin/bash
#PBS -l walltime=04:00:00
#PBS -l select=1:ncpus=10:mem=80gb
#PBS -N zMethAlign
#PBS -J 1-50
#--------------------------------------------
# 1. Initialise conda
#--------------------------------------------
eval "$(~/anaconda3/bin/conda shell.bash hook)"
#--------------------------------------------
# 2. Establish input/output streams
#--------------------------------------------
BAM_FILE=$(sed -n "${PBS_ARRAY_INDEX}p" /rds/general/user/hchown/projects/fisher-aspergillus-analysis/live/tutorial/assembly-annotation/scripts/bam_file_list.txt)
RUN_ID=$(echo "$BAM_FILE" | rev | cut -d"/" -f2 | rev)
FASTA_DIR=~/../ephemeral/
BAM_OUTDIR=~/../ephemeral/$RUN_ID/bam
#--------------------------------------------
# 3. Align BAM to genome
#--------------------------------------------
cd $TMPDIR
conda activate polish
mkdir -p $BAM_OUTDIR
~/../projects/fisher-local-software/live/dorado-1.3.0-linux-x64/bin/dorado aligner -t 10 $FASTA_DIR/$RUN_ID.ont.polished.fasta $BAM_FILE \
| samtools sort --threads 10 > $BAM_OUTDIR/$RUN_ID.$PBS_ARRAY_INDEX.aligned_reads.bam

```
Merging
samtools merge -o merged_output.bam input1.bam input2.bam input3.bam

Calling